{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 60,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/var/folders/1q/nqyt2x9s071fkn235hfq1p680000gn/T/ipykernel_29757/2078132498.py:5: DtypeWarning: Columns (0) have mixed types. Specify dtype option on import or set low_memory=False.\n",
      "  obs_gene_expr = pd.read_csv(gene_expr_bed, sep=\"\\t\", header=0, index_col='gene_id')\n"
     ]
    }
   ],
   "source": [
    "gene_expr_bed = \"/Users/sabrinami/Library/CloudStorage/Box-Box/imlab-data/data-Github/rat-genomic-analysis/Enformer-Predictions/data/Brain.rn7.expr.tpm.bed\"\n",
    "\n",
    "import pandas as pd\n",
    "import numpy as np\n",
    "obs_gene_expr = pd.read_csv(gene_expr_bed, sep=\"\\t\", header=0, index_col='gene_id')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>#chr</th>\n",
       "      <th>start</th>\n",
       "      <th>end</th>\n",
       "      <th>000789972A</th>\n",
       "      <th>0007899884</th>\n",
       "      <th>000789FF64</th>\n",
       "      <th>00079298B1</th>\n",
       "      <th>00079299B5</th>\n",
       "      <th>0007929A79</th>\n",
       "      <th>0007929B36</th>\n",
       "      <th>...</th>\n",
       "      <th>0007899761</th>\n",
       "      <th>0007899866</th>\n",
       "      <th>00078A00E6</th>\n",
       "      <th>0007929938</th>\n",
       "      <th>0007929956</th>\n",
       "      <th>00079FF7D6</th>\n",
       "      <th>0007899826</th>\n",
       "      <th>0007929917</th>\n",
       "      <th>00077E82E3</th>\n",
       "      <th>00078A18A7</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>gene_id</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>ENSRNOG00000065394</th>\n",
       "      <td>1</td>\n",
       "      <td>31563</td>\n",
       "      <td>31564</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSRNOG00000065356</th>\n",
       "      <td>1</td>\n",
       "      <td>41634</td>\n",
       "      <td>41635</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSRNOG00000070568</th>\n",
       "      <td>1</td>\n",
       "      <td>76833</td>\n",
       "      <td>76834</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.01</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSRNOG00000065706</th>\n",
       "      <td>1</td>\n",
       "      <td>184235</td>\n",
       "      <td>184236</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.18</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSRNOG00000069644</th>\n",
       "      <td>1</td>\n",
       "      <td>272252</td>\n",
       "      <td>272253</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows Ã— 343 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                   #chr   start     end  000789972A  0007899884  000789FF64  \\\n",
       "gene_id                                                                       \n",
       "ENSRNOG00000065394    1   31563   31564         0.0         0.0         0.0   \n",
       "ENSRNOG00000065356    1   41634   41635         0.0         0.0         0.0   \n",
       "ENSRNOG00000070568    1   76833   76834         0.0         0.0         0.0   \n",
       "ENSRNOG00000065706    1  184235  184236         0.0         0.0         0.0   \n",
       "ENSRNOG00000069644    1  272252  272253         0.0         0.0         0.0   \n",
       "\n",
       "                    00079298B1  00079299B5  0007929A79  0007929B36  ...  \\\n",
       "gene_id                                                             ...   \n",
       "ENSRNOG00000065394         0.0        0.00         0.0         0.0  ...   \n",
       "ENSRNOG00000065356         0.0        0.00         0.0         0.0  ...   \n",
       "ENSRNOG00000070568         0.0        0.00         0.0         0.0  ...   \n",
       "ENSRNOG00000065706         0.0        0.18         0.0         0.0  ...   \n",
       "ENSRNOG00000069644         0.0        0.00         0.0         0.0  ...   \n",
       "\n",
       "                    0007899761  0007899866  00078A00E6  0007929938  \\\n",
       "gene_id                                                              \n",
       "ENSRNOG00000065394         0.0         0.0         0.0         0.0   \n",
       "ENSRNOG00000065356         0.0         0.0         0.0         0.0   \n",
       "ENSRNOG00000070568         0.0         0.0         0.0         0.0   \n",
       "ENSRNOG00000065706         0.0         0.0         0.0         0.0   \n",
       "ENSRNOG00000069644         0.0         0.0         0.0         0.0   \n",
       "\n",
       "                    0007929956  00079FF7D6  0007899826  0007929917  \\\n",
       "gene_id                                                              \n",
       "ENSRNOG00000065394         0.0        0.00         0.0         0.0   \n",
       "ENSRNOG00000065356         0.0        0.00         0.0         0.0   \n",
       "ENSRNOG00000070568         0.0        0.01         0.0         0.0   \n",
       "ENSRNOG00000065706         0.0        0.00         0.0         0.0   \n",
       "ENSRNOG00000069644         0.0        0.00         0.0         0.0   \n",
       "\n",
       "                    00077E82E3  00078A18A7  \n",
       "gene_id                                     \n",
       "ENSRNOG00000065394         0.0         0.0  \n",
       "ENSRNOG00000065356         0.0         0.0  \n",
       "ENSRNOG00000070568         0.0         0.0  \n",
       "ENSRNOG00000065706         0.0         0.0  \n",
       "ENSRNOG00000069644         0.0         0.0  \n",
       "\n",
       "[5 rows x 343 columns]"
      ]
     },
     "execution_count": 4,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "\n",
    "obs_gene_expr.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 98,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "000789972A     7.53\n",
      "0007899884    11.35\n",
      "000789FF64     9.33\n",
      "00079298B1     9.88\n",
      "00079299B5      9.9\n",
      "              ...  \n",
      "00079FF7D6     7.63\n",
      "0007899826     2.83\n",
      "0007929917     8.29\n",
      "00077E82E3    10.05\n",
      "00078A18A7     6.73\n",
      "Name: ENSRNOG00000001342, Length: 340, dtype: object\n",
      "000789972A    44.43\n",
      "0007899884    50.12\n",
      "000789FF64    62.47\n",
      "00079298B1    64.97\n",
      "00079299B5    57.53\n",
      "              ...  \n",
      "00079FF7D6     53.3\n",
      "0007899826    52.88\n",
      "0007929917    54.19\n",
      "00077E82E3    74.91\n",
      "00078A18A7     50.0\n",
      "Name: ENSRNOG00000020624, Length: 340, dtype: object\n"
     ]
    }
   ],
   "source": [
    "gene_list = ['ENSRNOG00000001342', 'ENSRNOG00000020624']\n",
    "gene = gene_list[0]\n",
    "#obs_gene_expr[obs_gene_expr[\"gene_id\"].isin(gene_list)]\n",
    "\n",
    "pred_vs_obs = {}\n",
    "for gene in gene_list:\n",
    "    obs = obs_gene_expr.loc[gene][3:]\n",
    "    print(obs)\n",
    "    pred_vs_obs[gene] = pd.DataFrame({\"observed\": obs, \"predicted\": np.zeros(obs.shape)})"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 51,
   "metadata": {},
   "outputs": [],
   "source": [
    "def compute_tss_bin(output_window_coords_resized, tss_position, output_length=114688, bin_length=128, rc=False):\n",
    "    '''\n",
    "    Computes the bin index of the enformer output bin containing the TSS\n",
    "    Inputs:\n",
    "    output_window_coords_resized: Kipoiseq interval for the output window of Enformer, should already be resized to the model output window size\n",
    "    tss_position: absolute location of TSS\n",
    "    output_length: Length of output window in base pairs\n",
    "    bin_size: Length of output bins in base pairs\n",
    "    Outputs:\n",
    "    tss_bin: Index of bin containing TSS\n",
    "    '''\n",
    "    try:\n",
    "        relative_position = int(tss_position) - int(output_window_coords_resized.start)\n",
    "    except:\n",
    "    # print(tss_position)\n",
    "    # print(output_window_coords_resized.start)\n",
    "        sys.exit()\n",
    "\n",
    "    if relative_position < 0 or relative_position > output_length:\n",
    "        # print(tss_position, \"appears to be out of bounds for output window: \",output_window_coords_resized)\n",
    "        return None \n",
    "    else:\n",
    "        center = relative_position // bin_length\n",
    "\n",
    "        \n",
    "        if rc:\n",
    "            number_of_bins = int(output_length/bin_length)\n",
    "            new_center = number_of_bins - 1 - center\n",
    "            # print(new_center)\n",
    "            return (new_center-1,new_center,new_center+1)\n",
    "        else:\n",
    "            # print(center)\n",
    "            return (center-1,center,center+1)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/var/folders/1q/nqyt2x9s071fkn235hfq1p680000gn/T/ipykernel_29757/10523050.py:1: DtypeWarning: Columns (2) have mixed types. Specify dtype option on import or set low_memory=False.\n",
      "  annot_df = pd.read_csv(\"/Users/sabrinami/Library/CloudStorage/Box-Box/imlab-data/data-Github/rat-genomic-analysis/Enformer-Predictions/data/rn7.gene.txt\", sep=\"\\t\", index_col='geneId')\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>geneSymbol</th>\n",
       "      <th>chromosome</th>\n",
       "      <th>start</th>\n",
       "      <th>end</th>\n",
       "      <th>strand</th>\n",
       "      <th>tss</th>\n",
       "      <th>description</th>\n",
       "      <th>hasEqtl</th>\n",
       "      <th>expr_BLA</th>\n",
       "      <th>expr_Brain</th>\n",
       "      <th>...</th>\n",
       "      <th>sqtl_BLA</th>\n",
       "      <th>sqtl_Brain</th>\n",
       "      <th>sqtl_Eye</th>\n",
       "      <th>sqtl_IL</th>\n",
       "      <th>sqtl_LHb</th>\n",
       "      <th>sqtl_NAcc</th>\n",
       "      <th>sqtl_NAcc2</th>\n",
       "      <th>sqtl_OFC</th>\n",
       "      <th>sqtl_PL</th>\n",
       "      <th>sqtl_PL2</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>geneId</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>ENSRNOG00000066169</th>\n",
       "      <td>ENSRNOG00000066169</td>\n",
       "      <td>1</td>\n",
       "      <td>36112690</td>\n",
       "      <td>36122387</td>\n",
       "      <td>-</td>\n",
       "      <td>36122387</td>\n",
       "      <td>NaN</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>...</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSRNOG00000070168</th>\n",
       "      <td>Olr56</td>\n",
       "      <td>1</td>\n",
       "      <td>157231467</td>\n",
       "      <td>157232417</td>\n",
       "      <td>+</td>\n",
       "      <td>157231467</td>\n",
       "      <td>olfactory receptor family 51 subfamily F membe...</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>...</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSRNOG00000070901</th>\n",
       "      <td>Irgq</td>\n",
       "      <td>1</td>\n",
       "      <td>80123925</td>\n",
       "      <td>80131881</td>\n",
       "      <td>+</td>\n",
       "      <td>80123925</td>\n",
       "      <td>immunity-related GTPase Q</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>...</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSRNOG00000018029</th>\n",
       "      <td>Doc2g</td>\n",
       "      <td>1</td>\n",
       "      <td>201296342</td>\n",
       "      <td>201299865</td>\n",
       "      <td>+</td>\n",
       "      <td>201296342</td>\n",
       "      <td>double C2-like domains, gamma</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>...</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSRNOG00000031391</th>\n",
       "      <td>Ceacam16</td>\n",
       "      <td>1</td>\n",
       "      <td>79514975</td>\n",
       "      <td>79524871</td>\n",
       "      <td>-</td>\n",
       "      <td>79524871</td>\n",
       "      <td>CEA cell adhesion molecule 16, tectorial membr...</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>...</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows Ã— 68 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                            geneSymbol chromosome      start        end  \\\n",
       "geneId                                                                    \n",
       "ENSRNOG00000066169  ENSRNOG00000066169          1   36112690   36122387   \n",
       "ENSRNOG00000070168               Olr56          1  157231467  157232417   \n",
       "ENSRNOG00000070901                Irgq          1   80123925   80131881   \n",
       "ENSRNOG00000018029               Doc2g          1  201296342  201299865   \n",
       "ENSRNOG00000031391            Ceacam16          1   79514975   79524871   \n",
       "\n",
       "                   strand        tss  \\\n",
       "geneId                                 \n",
       "ENSRNOG00000066169      -   36122387   \n",
       "ENSRNOG00000070168      +  157231467   \n",
       "ENSRNOG00000070901      +   80123925   \n",
       "ENSRNOG00000018029      +  201296342   \n",
       "ENSRNOG00000031391      -   79524871   \n",
       "\n",
       "                                                          description  \\\n",
       "geneId                                                                  \n",
       "ENSRNOG00000066169                                                NaN   \n",
       "ENSRNOG00000070168  olfactory receptor family 51 subfamily F membe...   \n",
       "ENSRNOG00000070901                          immunity-related GTPase Q   \n",
       "ENSRNOG00000018029                      double C2-like domains, gamma   \n",
       "ENSRNOG00000031391  CEA cell adhesion molecule 16, tectorial membr...   \n",
       "\n",
       "                    hasEqtl  expr_BLA  expr_Brain  ...  sqtl_BLA  sqtl_Brain  \\\n",
       "geneId                                             ...                         \n",
       "ENSRNOG00000066169     True      True        True  ...     False       False   \n",
       "ENSRNOG00000070168     True     False        True  ...     False       False   \n",
       "ENSRNOG00000070901     True      True        True  ...     False       False   \n",
       "ENSRNOG00000018029    False      True        True  ...     False       False   \n",
       "ENSRNOG00000031391     True      True        True  ...     False       False   \n",
       "\n",
       "                    sqtl_Eye  sqtl_IL  sqtl_LHb  sqtl_NAcc  sqtl_NAcc2  \\\n",
       "geneId                                                                   \n",
       "ENSRNOG00000066169     False    False     False      False       False   \n",
       "ENSRNOG00000070168     False    False     False      False       False   \n",
       "ENSRNOG00000070901     False    False     False      False       False   \n",
       "ENSRNOG00000018029     False    False     False      False       False   \n",
       "ENSRNOG00000031391     False    False     False      False       False   \n",
       "\n",
       "                    sqtl_OFC  sqtl_PL  sqtl_PL2  \n",
       "geneId                                           \n",
       "ENSRNOG00000066169     False    False     False  \n",
       "ENSRNOG00000070168     False    False     False  \n",
       "ENSRNOG00000070901     False    False     False  \n",
       "ENSRNOG00000018029     False    False     False  \n",
       "ENSRNOG00000031391     False    False     False  \n",
       "\n",
       "[5 rows x 68 columns]"
      ]
     },
     "execution_count": 20,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "annot_df = pd.read_csv(\"/Users/sabrinami/Library/CloudStorage/Box-Box/imlab-data/data-Github/rat-genomic-analysis/Enformer-Predictions/data/rn7.gene.txt\", sep=\"\\t\", index_col='geneId')\n",
    "annot_df.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 108,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "           observed  predicted\n",
      "000789972A     7.53        0.0\n",
      "0007899884    11.35        0.0\n",
      "000789FF64     9.33        0.0\n",
      "00079298B1     9.88        0.0\n",
      "00079299B5      9.9        0.0\n",
      "...             ...        ...\n",
      "00079FF7D6     7.63        0.0\n",
      "0007899826     2.83        0.0\n",
      "0007929917     8.29        0.0\n",
      "00077E82E3    10.05        0.0\n",
      "00078A18A7     6.73        0.0\n",
      "\n",
      "[340 rows x 2 columns]\n",
      "           observed  predicted\n",
      "000789972A    44.43        0.0\n",
      "0007899884    50.12        0.0\n",
      "000789FF64    62.47        0.0\n",
      "00079298B1    64.97        0.0\n",
      "00079299B5    57.53        0.0\n",
      "...             ...        ...\n",
      "00079FF7D6     53.3        0.0\n",
      "0007899826    52.88        0.0\n",
      "0007929917    54.19        0.0\n",
      "00077E82E3    74.91        0.0\n",
      "00078A18A7     50.0        0.0\n",
      "\n",
      "[340 rows x 2 columns]\n"
     ]
    }
   ],
   "source": [
    "import kipoiseq # for manipulating fasta files\n",
    "from kipoiseq import Interval # same as above, really\n",
    "import sys\n",
    "import h5py\n",
    "predictions_dir = \"/Users/sabrinami/Library/CloudStorage/Box-Box/imlab-data/data-Github/rat-genomic-analysis/Enformer-Predictions/predictions\"\n",
    "\n",
    "for gene in gene_list:\n",
    "    gene_annot = annot_df.loc[gene]\n",
    "    interval_obj = Interval(\"chr\" + str(gene_annot['chromosome']), gene_annot['start'], gene_annot['end'])\n",
    "    tss_bins = compute_tss_bin(interval_obj.resize(114688), gene_annot['tss'])\n",
    "    interval = f\"chr{gene_annot['chromosome']}_{gene_annot['start']}_{gene_annot['end']}\"\n",
    "    for individual in pred_vs_obs[gene].index:\n",
    "        individual = \"00079FB742\"\n",
    "        predictions = h5py.File(f\"{predictions_dir}/{individual}/haplotype1/{interval}_predictions.h5\", \"r\")\n",
    "        matrix = predictions[interval][()]\n",
    "        pred_expr = np.average(matrix[tss_bins, 4980])\n",
    "        pred_vs_obs[gene].loc[individual, \"predicted\"] = pred_expr\n",
    "        break"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "base",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.9.16"
  },
  "orig_nbformat": 4
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
