{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "---\n",
    "title: Calculating Br rat haplotype probabilities from HS founders\n",
    "date: 11/8/2023\n",
    "author: Sabrina\n",
    "---\n",
    "\n",
    "I am using Dan Munro's scripts to compute probabilities across the 8 founders per locus per individual. His code uses the R qtl2 package.\n",
    "\n",
    "Github: [https://github.com/daniel-munro/qtl2-founder-haps/tree/main](https://github.com/daniel-munro/qtl2-founder-haps/tree/main)\n",
    "\n",
    "The `genetic_map` subdirectory contains genetic mapping files from the older build rn6. I downloaded rn7 genetic markers from his preprint, [A revamped rat reference genome improves the discovery of genetic diversity in laboratory rats](https://www.biorxiv.org/content/10.1101/2023.04.13.536694v2.article-info), [Supplementary Table S2](https://www.biorxiv.org/content/biorxiv/early/2023/09/29/2023.04.13.536694/DC2/embed/media-2.xlsx?download=true).\n",
    "\n",
    "I first saved Table S2 as a CSV, then split it by chromosome."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>CONTIG</th>\n",
       "      <th>POS</th>\n",
       "      <th>CHR</th>\n",
       "      <th>cM</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>chr1</td>\n",
       "      <td>257394712</td>\n",
       "      <td>1</td>\n",
       "      <td>0.0000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>chr1</td>\n",
       "      <td>257401401</td>\n",
       "      <td>1</td>\n",
       "      <td>0.0000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>chr1</td>\n",
       "      <td>257385252</td>\n",
       "      <td>1</td>\n",
       "      <td>0.4775</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>chr1</td>\n",
       "      <td>257412479</td>\n",
       "      <td>1</td>\n",
       "      <td>0.6100</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>chr1</td>\n",
       "      <td>257366293</td>\n",
       "      <td>1</td>\n",
       "      <td>0.6100</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>150830</th>\n",
       "      <td>chr20</td>\n",
       "      <td>51679712</td>\n",
       "      <td>20</td>\n",
       "      <td>66.0760</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>150831</th>\n",
       "      <td>chr20</td>\n",
       "      <td>51749221</td>\n",
       "      <td>20</td>\n",
       "      <td>66.0760</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>150832</th>\n",
       "      <td>chr20</td>\n",
       "      <td>51731733</td>\n",
       "      <td>20</td>\n",
       "      <td>66.0760</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>150833</th>\n",
       "      <td>chr20</td>\n",
       "      <td>51651864</td>\n",
       "      <td>20</td>\n",
       "      <td>66.0760</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>150834</th>\n",
       "      <td>chr20</td>\n",
       "      <td>51685591</td>\n",
       "      <td>20</td>\n",
       "      <td>66.0760</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>150835 rows Ã— 4 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "       CONTIG        POS  CHR       cM\n",
       "0        chr1  257394712    1   0.0000\n",
       "1        chr1  257401401    1   0.0000\n",
       "2        chr1  257385252    1   0.4775\n",
       "3        chr1  257412479    1   0.6100\n",
       "4        chr1  257366293    1   0.6100\n",
       "...       ...        ...  ...      ...\n",
       "150830  chr20   51679712   20  66.0760\n",
       "150831  chr20   51749221   20  66.0760\n",
       "150832  chr20   51731733   20  66.0760\n",
       "150833  chr20   51651864   20  66.0760\n",
       "150834  chr20   51685591   20  66.0760\n",
       "\n",
       "[150835 rows x 4 columns]"
      ]
     },
     "execution_count": 1,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "import pandas as pd\n",
    "import gzip\n",
    "\n",
    "# Read the original CSV file\n",
    "df = pd.read_csv('/Users/sabrinami/Desktop/MAP.csv')\n",
    "\n",
    "# Drop the 'CONTIG' column\n",
    "# df = df.drop(columns=['CONTIG'])"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.11.6"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
