---
title: Preparing HS founder rat genotypes for Enformer pipeline
date: 9/12/2023
author: Sabrina Mi
engine: knitr
---

I downloaded the original VCF from TSCC to `/home/s1mi/enformer_rat_data/genotypes/Palmer_HS_founders_mRatBN7_2.vcf.gz`.




### Convert NCBI RefSeq ID to chromosome numbers

1. Download assembly report for NCBI names to chromosome mapping

```
wget https://ftp.ncbi.nlm.nih.gov/genomes/all/GCF/015/227/675/GCF_015227675.2_mRatBN7.2/GCF_015227675.2_mRatBN7.2_assembly_report.txt -O /Users/sabrinami/Downloads/mRatBN7.2_assembly_report.txt
```

2. Write chromosome annotation file for bcftools

```
grep -e '^[^#]' /Users/sabrinami/Downloads/mRatBN7.2_assembly_report.txt | awk '{ print $7, "chr" $1 }' > /Users/sabrinami/Downloads/mRatBN7.2_chromosome_names.txt

```

There are extra annotations than needed, later on we will filter to chromosomes 1 through 20 and X.

3. Annotate VCF

```
cd /Users/sabrinami/Desktop/Sabrina/2022-23/tutorials/enformer_pipeline_test/rn7_data
bcftools annotate \
  --rename-chrs /Users/sabrinami/Downloads/mRatBN7.2_chromosome_names.txt \
  --threads 10 \
  -o Palmer_HS_founders_mRatBN7_annot.vcf \
  -Ov Palmer_HS_founders_mRatBN7_2.vcf.gz

```

### Replace Missing Genotypes with Reference Alleles

```
sed 's/\.\/\./0\/0/g' Palmer_HS_founders_mRatBN7_annot.vcf > Palmer_HS_founders_mRatBN7_no_missing.vcf

```

### Filter out Indels

```
bcftools view --types snps Palmer_HS_founders_mRatBN7_no_missing.vcf -o Palmer_HS_founders_mRatBN7_snps_only.vcf -Ov
```

Some sites have deletions notated by `*` in the ALT field in addition to an allele:

```{bash}
cd /Users/sabrinami/Desktop/Sabrina/2022-23/tutorials/enformer_pipeline_test/rn7_data
grep '\*' Palmer_HS_founders_mRatBN7_snps_only.vcf | head -n 5

```

We want to retain data for those samples with the alternative allele, but drop samples with the deletions. This is done in two steps, by first splitting the ALT column and then filtering out the deletion in the newly created record:

```

bcftools norm -m +snps Palmer_HS_founders_mRatBN7_snps_only.vcf | bcftools view -e 'ALT="*"' -o Palmer_HS_founders_mRatBN7_biallelic_snps.vcf.gz -Oz
bcftools index -t Palmer_HS_founders_mRatBN7_biallelic_snps.vcf.gz
```


### Split VCFs

First by chromosome number:

```
mkdir FounderVCFs
# Split VCF by chromosome
vcf_in=Palmer_HS_founders_mRatBN7_biallelic_snps.vcf.gz

vcf_out_prefix=FounderVCFs/chr

for i in {1..20}
do
    echo "Working on chromosome ${i}..."
    bcftools view --regions chr${i} ${vcf_in} -o ${vcf_out_prefix}${i}.vcf.gz -Oz
done


# Index VCFs
for i in {1..20}
do
    echo "Indexing chromosome ${i}..."
    bcftools index -t FounderVCFs/chr${i}.vcf.gz
done
```

Then separately for X chromosome:

```
bcftools view ${vcf_in} --regions chrX -o ${vcf_out_prefix}X.vcf.gz -Oz
bcftools index -t FounderVCFs/chrX.vcf.gz
```