---
title: Preparing HS founder rat genotypes for Enformer pipeline
date: 9/12/2023
author: Sabrina Mi
engine: knitr
---

I downloaded the original VCF from TSCC to `/home/s1mi/enformer_rat_data/genotypes/Palmer_HS_founders_mRatBN7_2.vcf.gz`.

So far, I've filtered out indels and replaced samples with missing genotype info with reference alleles. The current issue preventing me from running the enformer pipeline are the bi-allelic variants with a missing an alternative allele:

```{bash}
cd /Users/sabrinami/Desktop/Sabrina/2022-23/tutorials/enformer_pipeline_test/rn7_data
zless Palmer_HS_founders_mRatBN7_snps_only.vcf.gz | grep '\*' | head -n 5

```

I had a lot of conversations with people on handling them and trying to cross-reference outbred and inbred rat variants. Worst case, I will assign the missing allele to the other recorded allele. Den thinks he can fill in some of the missing `*` alleles with long reads from the HS founders. 


### Convert NCBI RefSeq ID to chromosome numbers

1. Download assembly report for NCBI names to chromosome mapping

```
wget https://ftp.ncbi.nlm.nih.gov/genomes/all/GCF/015/227/675/GCF_015227675.2_mRatBN7.2/GCF_015227675.2_mRatBN7.2_assembly_report.txt -O /Users/sabrinami/Downloads/mRatBN7.2_assembly_report.txt
```

2. Write chromosome annotation file for bcftools

```
grep -e '^[^#]' /Users/sabrinami/Downloads/mRatBN7.2_assembly_report.txt | awk '{ print $7, "chr" $1 }' > /Users/sabrinami/Downloads/mRatBN7.2_chromosome_names.txt

```

There are extra annotations than needed, later on we will filter to chromosomes 1 through 20 and X.

3. Annotate VCF

```
cd /Users/sabrinami/Desktop/Sabrina/2022-23/tutorials/enformer_pipeline_test/rn7_data
bcftools annotate \
  --rename-chrs /Users/sabrinami/Downloads/mRatBN7.2_chromosome_names.txt \
  --threads 10 \
  -o Palmer_HS_founders_mRatBN7_annot.vcf.gz \
  -Oz Palmer_HS_founders_mRatBN7_2.vcf.gz

```

### Replace Missing Genotypes with Reference Alleles

```
gunzip Palmer_HS_founders_mRatBN7_annot.vcf.gz
sed 's/\.\/\./0\/0/g' Palmer_HS_founders_mRatBN7_annot.vcf > Palmer_HS_founders_mRatBN7_no_missing.vcf
bgzip Palmer_HS_founders_mRatBN7_no_missing.vcf
bcftools index -t Palmer_HS_founders_mRatBN7_no_missing.vcf.gz

```

### Filter out Indels

```
bcftools view --types snps Palmer_HS_founders_mRatBN7_no_missing.vcf.gz -o Palmer_HS_founders_mRatBN7_snps_only.vcf.gz -Oz
bcftools index -t Palmer_HS_founders_mRatBN7_snps_only.vcf.gz
```

### Split VCFs

First by chromosome number:

```
mkdir FounderVCFs
# Split VCF by chromosome
vcf_in=Palmer_HS_founders_mRatBN7_snps_only.vcf.gz

vcf_out_prefix=FounderVCFs/chr

for i in {1..20}
do
    echo "Working on chromosome ${i}..."
    bcftools view --regions chr${i} ${vcf_in} -o ${vcf_out_prefix}${i}.vcf.gz -Oz
done


# Index VCFs
for i in {1..20}
do
    echo "Indexing chromosome ${i}..."
    bcftools index -t FounderVCFs/chr${i}.vcf.gz
done
```

Then separately for X chromosome:

```
bcftools view ${vcf_in} --regions chrX -o ${vcf_out_prefix}X.vcf.gz -Oz
bcftools index -t FounderVCFs/chrX.vcf.gz
```