{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "---\n",
    "title: Splitting chr20 Brain rats into batches for qtl2 analysis\n",
    "author: Sabrina Mi\n",
    "date: 11/19/2023\n",
    "---\n",
    "\n",
    "We'll split the 340 Br rats into batches of 10. The goal is to do this without creating a new VCF for each batchThe following code subsets the chr20 input files, specifically geno.csv and covar.csv files, that were generated with [qtl2-founder-haps code](https://sabrina-dl.hakyimlab.org/posts/2023-11-14-calculating-genotype-probabilities-by-chromosome/). I set up by copying over the other input files that should stay the same.\n",
    "\n",
    "```\n",
    "mkdir tmp-chr20-qtl2-founder-haps\n",
    "mkdir chr20-qtl2-outputs\n",
    "cp chr20-qtl2-founder-haps/control.yaml tmp-chr20-qtl2-founder-haps/control.yaml\n",
    "cp chr20-qtl2-founder-haps/founder_geno.csv tmp-chr20-qtl2-founder-haps/founder_geno.csv\n",
    "cp chr20-qtl2-founder-haps/gmap.csv tmp-chr20-qtl2-founder-haps/gmap.csv\n",
    "cp chr20-qtl2-founder-haps/pmap.csv tmp-chr20-qtl2-founder-haps/pmap.csv\n",
    "```"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "import time\n",
    "import subprocess\n",
    "data_dir = '/home/s1mi/chr20-qtl2-founder-haps'\n",
    "tmp_dir = '/home/s1mi/tmp-chr20-qtl2-founder-haps'\n",
    "output_dir = '/home/s1mi/chr20-qtl2-outputs'\n",
    "geno = pd.read_csv(f'{data_dir}/geno.csv', index_col='id')\n",
    "batch_size = 10\n",
    "n_batches = len(geno.columns) // batch_size"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [],
   "source": [
    "def qtl_command(output_file, n_cores=1):\n",
    "    cmd = (\n",
    "        'library(qtl2); '\n",
    "        f'cross <- read_cross2(\"{tmp_dir}/control.yaml\"); '\n",
    "        f'pr <- calc_genoprob(cross, error_prob = 0.01, cores = {n_cores}); '\n",
    "        f'pr <- genoprob_to_alleleprob(pr); saveRDS(pr, \"{output_file}\")'\n",
    "    )\n",
    "    return cmd"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "R version 4.3.2 (2023-10-31) -- \"Eye Holes\"\n",
      "Copyright (C) 2023 The R Foundation for Statistical Computing\n",
      "Platform: x86_64-conda-linux-gnu (64-bit)\n",
      "\n",
      "R is free software and comes with ABSOLUTELY NO WARRANTY.\n",
      "You are welcome to redistribute it under certain conditions.\n",
      "Type 'license()' or 'licence()' for distribution details.\n",
      "\n",
      "  Natural language support but running in an English locale\n",
      "\n",
      "R is a collaborative project with many contributors.\n",
      "Type 'contributors()' for more information and\n",
      "'citation()' on how to cite R or R packages in publications.\n",
      "\n",
      "Type 'demo()' for some demos, 'help()' for on-line help, or\n",
      "'help.start()' for an HTML browser interface to help.\n",
      "Type 'q()' to quit R.\n",
      "\n",
      "> library(qtl2); cross <- read_cross2(\"/home/s1mi/tmp-chr20-qtl2-founder-haps/control.yaml\"); pr <- calc_genoprob(cross, error_prob = 0.01, cores = 2); pr <- genoprob_to_alleleprob(pr); saveRDS(pr, \"/home/s1mi/chr20-qtl2-outputs/batch0_prob.rds\")\n",
      "> \n",
      "> \n",
      "Batch: 1 ... 5.516161077900324 minutes\n",
      "\n",
      "R version 4.3.2 (2023-10-31) -- \"Eye Holes\"\n",
      "Copyright (C) 2023 The R Foundation for Statistical Computing\n",
      "Platform: x86_64-conda-linux-gnu (64-bit)\n",
      "\n",
      "R is free software and comes with ABSOLUTELY NO WARRANTY.\n",
      "You are welcome to redistribute it under certain conditions.\n",
      "Type 'license()' or 'licence()' for distribution details.\n",
      "\n",
      "  Natural language support but running in an English locale\n",
      "\n",
      "R is a collaborative project with many contributors.\n",
      "Type 'contributors()' for more information and\n",
      "'citation()' on how to cite R or R packages in publications.\n",
      "\n",
      "Type 'demo()' for some demos, 'help()' for on-line help, or\n",
      "'help.start()' for an HTML browser interface to help.\n",
      "Type 'q()' to quit R.\n",
      "\n",
      "> library(qtl2); cross <- read_cross2(\"/home/s1mi/tmp-chr20-qtl2-founder-haps/control.yaml\"); pr <- calc_genoprob(cross, error_prob = 0.01, cores = 2); pr <- genoprob_to_alleleprob(pr); saveRDS(pr, \"/home/s1mi/chr20-qtl2-outputs/batch1_prob.rds\")\n",
      "> \n",
      "> \n",
      "Batch: 2 ... 5.188156945933588 minutes\n",
      "\n",
      "R version 4.3.2 (2023-10-31) -- \"Eye Holes\"\n",
      "Copyright (C) 2023 The R Foundation for Statistical Computing\n",
      "Platform: x86_64-conda-linux-gnu (64-bit)\n",
      "\n",
      "R is free software and comes with ABSOLUTELY NO WARRANTY.\n",
      "You are welcome to redistribute it under certain conditions.\n",
      "Type 'license()' or 'licence()' for distribution details.\n",
      "\n",
      "  Natural language support but running in an English locale\n",
      "\n",
      "R is a collaborative project with many contributors.\n",
      "Type 'contributors()' for more information and\n",
      "'citation()' on how to cite R or R packages in publications.\n",
      "\n",
      "Type 'demo()' for some demos, 'help()' for on-line help, or\n",
      "'help.start()' for an HTML browser interface to help.\n",
      "Type 'q()' to quit R.\n",
      "\n",
      "> library(qtl2); cross <- read_cross2(\"/home/s1mi/tmp-chr20-qtl2-founder-haps/control.yaml\"); pr <- calc_genoprob(cross, error_prob = 0.01, cores = 2); pr <- genoprob_to_alleleprob(pr); saveRDS(pr, \"/home/s1mi/chr20-qtl2-outputs/batch2_prob.rds\")\n",
      "> \n",
      "> \n",
      "Batch: 3 ... 5.320735947232849 minutes\n",
      "\n",
      "R version 4.3.2 (2023-10-31) -- \"Eye Holes\"\n",
      "Copyright (C) 2023 The R Foundation for Statistical Computing\n",
      "Platform: x86_64-conda-linux-gnu (64-bit)\n",
      "\n",
      "R is free software and comes with ABSOLUTELY NO WARRANTY.\n",
      "You are welcome to redistribute it under certain conditions.\n",
      "Type 'license()' or 'licence()' for distribution details.\n",
      "\n",
      "  Natural language support but running in an English locale\n",
      "\n",
      "R is a collaborative project with many contributors.\n",
      "Type 'contributors()' for more information and\n",
      "'citation()' on how to cite R or R packages in publications.\n",
      "\n",
      "Type 'demo()' for some demos, 'help()' for on-line help, or\n",
      "'help.start()' for an HTML browser interface to help.\n",
      "Type 'q()' to quit R.\n",
      "\n",
      "> library(qtl2); cross <- read_cross2(\"/home/s1mi/tmp-chr20-qtl2-founder-haps/control.yaml\"); pr <- calc_genoprob(cross, error_prob = 0.01, cores = 2); pr <- genoprob_to_alleleprob(pr); saveRDS(pr, \"/home/s1mi/chr20-qtl2-outputs/batch3_prob.rds\")\n",
      "> \n",
      "> \n",
      "Batch: 4 ... 5.9591666307009294 minutes\n",
      "\n",
      "R version 4.3.2 (2023-10-31) -- \"Eye Holes\"\n",
      "Copyright (C) 2023 The R Foundation for Statistical Computing\n",
      "Platform: x86_64-conda-linux-gnu (64-bit)\n",
      "\n",
      "R is free software and comes with ABSOLUTELY NO WARRANTY.\n",
      "You are welcome to redistribute it under certain conditions.\n",
      "Type 'license()' or 'licence()' for distribution details.\n",
      "\n",
      "  Natural language support but running in an English locale\n",
      "\n",
      "R is a collaborative project with many contributors.\n",
      "Type 'contributors()' for more information and\n",
      "'citation()' on how to cite R or R packages in publications.\n",
      "\n",
      "Type 'demo()' for some demos, 'help()' for on-line help, or\n",
      "'help.start()' for an HTML browser interface to help.\n",
      "Type 'q()' to quit R.\n",
      "\n",
      "> library(qtl2); cross <- read_cross2(\"/home/s1mi/tmp-chr20-qtl2-founder-haps/control.yaml\"); pr <- calc_genoprob(cross, error_prob = 0.01, cores = 2); pr <- genoprob_to_alleleprob(pr); saveRDS(pr, \"/home/s1mi/chr20-qtl2-outputs/batch4_prob.rds\")\n",
      "> \n",
      "> \n",
      "Batch: 5 ... 5.304590424666336 minutes\n",
      "\n",
      "R version 4.3.2 (2023-10-31) -- \"Eye Holes\"\n",
      "Copyright (C) 2023 The R Foundation for Statistical Computing\n",
      "Platform: x86_64-conda-linux-gnu (64-bit)\n",
      "\n",
      "R is free software and comes with ABSOLUTELY NO WARRANTY.\n",
      "You are welcome to redistribute it under certain conditions.\n",
      "Type 'license()' or 'licence()' for distribution details.\n",
      "\n",
      "  Natural language support but running in an English locale\n",
      "\n",
      "R is a collaborative project with many contributors.\n",
      "Type 'contributors()' for more information and\n",
      "'citation()' on how to cite R or R packages in publications.\n",
      "\n",
      "Type 'demo()' for some demos, 'help()' for on-line help, or\n",
      "'help.start()' for an HTML browser interface to help.\n",
      "Type 'q()' to quit R.\n",
      "\n",
      "> library(qtl2); cross <- read_cross2(\"/home/s1mi/tmp-chr20-qtl2-founder-haps/control.yaml\"); pr <- calc_genoprob(cross, error_prob = 0.01, cores = 2); pr <- genoprob_to_alleleprob(pr); saveRDS(pr, \"/home/s1mi/chr20-qtl2-outputs/batch5_prob.rds\")\n",
      "> \n",
      "> \n",
      "Batch: 6 ... 5.208924225267643 minutes\n",
      "\n",
      "R version 4.3.2 (2023-10-31) -- \"Eye Holes\"\n",
      "Copyright (C) 2023 The R Foundation for Statistical Computing\n",
      "Platform: x86_64-conda-linux-gnu (64-bit)\n",
      "\n",
      "R is free software and comes with ABSOLUTELY NO WARRANTY.\n",
      "You are welcome to redistribute it under certain conditions.\n",
      "Type 'license()' or 'licence()' for distribution details.\n",
      "\n",
      "  Natural language support but running in an English locale\n",
      "\n",
      "R is a collaborative project with many contributors.\n",
      "Type 'contributors()' for more information and\n",
      "'citation()' on how to cite R or R packages in publications.\n",
      "\n",
      "Type 'demo()' for some demos, 'help()' for on-line help, or\n",
      "'help.start()' for an HTML browser interface to help.\n",
      "Type 'q()' to quit R.\n",
      "\n",
      "> library(qtl2); cross <- read_cross2(\"/home/s1mi/tmp-chr20-qtl2-founder-haps/control.yaml\"); pr <- calc_genoprob(cross, error_prob = 0.01, cores = 2); pr <- genoprob_to_alleleprob(pr); saveRDS(pr, \"/home/s1mi/chr20-qtl2-outputs/batch6_prob.rds\")\n",
      "> \n",
      "> \n",
      "Batch: 7 ... 5.362514834082686 minutes\n",
      "\n",
      "R version 4.3.2 (2023-10-31) -- \"Eye Holes\"\n",
      "Copyright (C) 2023 The R Foundation for Statistical Computing\n",
      "Platform: x86_64-conda-linux-gnu (64-bit)\n",
      "\n",
      "R is free software and comes with ABSOLUTELY NO WARRANTY.\n",
      "You are welcome to redistribute it under certain conditions.\n",
      "Type 'license()' or 'licence()' for distribution details.\n",
      "\n",
      "  Natural language support but running in an English locale\n",
      "\n",
      "R is a collaborative project with many contributors.\n",
      "Type 'contributors()' for more information and\n",
      "'citation()' on how to cite R or R packages in publications.\n",
      "\n",
      "Type 'demo()' for some demos, 'help()' for on-line help, or\n",
      "'help.start()' for an HTML browser interface to help.\n",
      "Type 'q()' to quit R.\n",
      "\n",
      "> library(qtl2); cross <- read_cross2(\"/home/s1mi/tmp-chr20-qtl2-founder-haps/control.yaml\"); pr <- calc_genoprob(cross, error_prob = 0.01, cores = 2); pr <- genoprob_to_alleleprob(pr); saveRDS(pr, \"/home/s1mi/chr20-qtl2-outputs/batch7_prob.rds\")\n",
      "> \n",
      "> \n",
      "Batch: 8 ... 5.318476139800623 minutes\n",
      "\n",
      "R version 4.3.2 (2023-10-31) -- \"Eye Holes\"\n",
      "Copyright (C) 2023 The R Foundation for Statistical Computing\n",
      "Platform: x86_64-conda-linux-gnu (64-bit)\n",
      "\n",
      "R is free software and comes with ABSOLUTELY NO WARRANTY.\n",
      "You are welcome to redistribute it under certain conditions.\n",
      "Type 'license()' or 'licence()' for distribution details.\n",
      "\n",
      "  Natural language support but running in an English locale\n",
      "\n",
      "R is a collaborative project with many contributors.\n",
      "Type 'contributors()' for more information and\n",
      "'citation()' on how to cite R or R packages in publications.\n",
      "\n",
      "Type 'demo()' for some demos, 'help()' for on-line help, or\n",
      "'help.start()' for an HTML browser interface to help.\n",
      "Type 'q()' to quit R.\n",
      "\n",
      "> library(qtl2); cross <- read_cross2(\"/home/s1mi/tmp-chr20-qtl2-founder-haps/control.yaml\"); pr <- calc_genoprob(cross, error_prob = 0.01, cores = 2); pr <- genoprob_to_alleleprob(pr); saveRDS(pr, \"/home/s1mi/chr20-qtl2-outputs/batch8_prob.rds\")\n",
      "> \n",
      "> \n",
      "Batch: 9 ... 5.43069331633354 minutes\n",
      "\n",
      "R version 4.3.2 (2023-10-31) -- \"Eye Holes\"\n",
      "Copyright (C) 2023 The R Foundation for Statistical Computing\n",
      "Platform: x86_64-conda-linux-gnu (64-bit)\n",
      "\n",
      "R is free software and comes with ABSOLUTELY NO WARRANTY.\n",
      "You are welcome to redistribute it under certain conditions.\n",
      "Type 'license()' or 'licence()' for distribution details.\n",
      "\n",
      "  Natural language support but running in an English locale\n",
      "\n",
      "R is a collaborative project with many contributors.\n",
      "Type 'contributors()' for more information and\n",
      "'citation()' on how to cite R or R packages in publications.\n",
      "\n",
      "Type 'demo()' for some demos, 'help()' for on-line help, or\n",
      "'help.start()' for an HTML browser interface to help.\n",
      "Type 'q()' to quit R.\n",
      "\n",
      "> library(qtl2); cross <- read_cross2(\"/home/s1mi/tmp-chr20-qtl2-founder-haps/control.yaml\"); pr <- calc_genoprob(cross, error_prob = 0.01, cores = 2); pr <- genoprob_to_alleleprob(pr); saveRDS(pr, \"/home/s1mi/chr20-qtl2-outputs/batch9_prob.rds\")\n",
      "> \n",
      "> \n",
      "Batch: 10 ... 5.109263749249901 minutes\n",
      "\n",
      "R version 4.3.2 (2023-10-31) -- \"Eye Holes\"\n",
      "Copyright (C) 2023 The R Foundation for Statistical Computing\n",
      "Platform: x86_64-conda-linux-gnu (64-bit)\n",
      "\n",
      "R is free software and comes with ABSOLUTELY NO WARRANTY.\n",
      "You are welcome to redistribute it under certain conditions.\n",
      "Type 'license()' or 'licence()' for distribution details.\n",
      "\n",
      "  Natural language support but running in an English locale\n",
      "\n",
      "R is a collaborative project with many contributors.\n",
      "Type 'contributors()' for more information and\n",
      "'citation()' on how to cite R or R packages in publications.\n",
      "\n",
      "Type 'demo()' for some demos, 'help()' for on-line help, or\n",
      "'help.start()' for an HTML browser interface to help.\n",
      "Type 'q()' to quit R.\n",
      "\n",
      "> library(qtl2); cross <- read_cross2(\"/home/s1mi/tmp-chr20-qtl2-founder-haps/control.yaml\"); pr <- calc_genoprob(cross, error_prob = 0.01, cores = 2); pr <- genoprob_to_alleleprob(pr); saveRDS(pr, \"/home/s1mi/chr20-qtl2-outputs/batch10_prob.rds\")\n",
      "> \n",
      "> \n",
      "Batch: 11 ... 5.2156281993336355 minutes\n",
      "\n",
      "R version 4.3.2 (2023-10-31) -- \"Eye Holes\"\n",
      "Copyright (C) 2023 The R Foundation for Statistical Computing\n",
      "Platform: x86_64-conda-linux-gnu (64-bit)\n",
      "\n",
      "R is free software and comes with ABSOLUTELY NO WARRANTY.\n",
      "You are welcome to redistribute it under certain conditions.\n",
      "Type 'license()' or 'licence()' for distribution details.\n",
      "\n",
      "  Natural language support but running in an English locale\n",
      "\n",
      "R is a collaborative project with many contributors.\n",
      "Type 'contributors()' for more information and\n",
      "'citation()' on how to cite R or R packages in publications.\n",
      "\n",
      "Type 'demo()' for some demos, 'help()' for on-line help, or\n",
      "'help.start()' for an HTML browser interface to help.\n",
      "Type 'q()' to quit R.\n",
      "\n",
      "> library(qtl2); cross <- read_cross2(\"/home/s1mi/tmp-chr20-qtl2-founder-haps/control.yaml\"); pr <- calc_genoprob(cross, error_prob = 0.01, cores = 2); pr <- genoprob_to_alleleprob(pr); saveRDS(pr, \"/home/s1mi/chr20-qtl2-outputs/batch11_prob.rds\")\n",
      "> \n",
      "> \n",
      "Batch: 12 ... 5.699721770366887 minutes\n",
      "\n",
      "R version 4.3.2 (2023-10-31) -- \"Eye Holes\"\n",
      "Copyright (C) 2023 The R Foundation for Statistical Computing\n",
      "Platform: x86_64-conda-linux-gnu (64-bit)\n",
      "\n",
      "R is free software and comes with ABSOLUTELY NO WARRANTY.\n",
      "You are welcome to redistribute it under certain conditions.\n",
      "Type 'license()' or 'licence()' for distribution details.\n",
      "\n",
      "  Natural language support but running in an English locale\n",
      "\n",
      "R is a collaborative project with many contributors.\n",
      "Type 'contributors()' for more information and\n",
      "'citation()' on how to cite R or R packages in publications.\n",
      "\n",
      "Type 'demo()' for some demos, 'help()' for on-line help, or\n",
      "'help.start()' for an HTML browser interface to help.\n",
      "Type 'q()' to quit R.\n",
      "\n",
      "> library(qtl2); cross <- read_cross2(\"/home/s1mi/tmp-chr20-qtl2-founder-haps/control.yaml\"); pr <- calc_genoprob(cross, error_prob = 0.01, cores = 2); pr <- genoprob_to_alleleprob(pr); saveRDS(pr, \"/home/s1mi/chr20-qtl2-outputs/batch12_prob.rds\")\n",
      "> \n",
      "> \n",
      "Batch: 13 ... 5.225824187783291 minutes\n",
      "\n",
      "R version 4.3.2 (2023-10-31) -- \"Eye Holes\"\n",
      "Copyright (C) 2023 The R Foundation for Statistical Computing\n",
      "Platform: x86_64-conda-linux-gnu (64-bit)\n",
      "\n",
      "R is free software and comes with ABSOLUTELY NO WARRANTY.\n",
      "You are welcome to redistribute it under certain conditions.\n",
      "Type 'license()' or 'licence()' for distribution details.\n",
      "\n",
      "  Natural language support but running in an English locale\n",
      "\n",
      "R is a collaborative project with many contributors.\n",
      "Type 'contributors()' for more information and\n",
      "'citation()' on how to cite R or R packages in publications.\n",
      "\n",
      "Type 'demo()' for some demos, 'help()' for on-line help, or\n",
      "'help.start()' for an HTML browser interface to help.\n",
      "Type 'q()' to quit R.\n",
      "\n",
      "> library(qtl2); cross <- read_cross2(\"/home/s1mi/tmp-chr20-qtl2-founder-haps/control.yaml\"); pr <- calc_genoprob(cross, error_prob = 0.01, cores = 2); pr <- genoprob_to_alleleprob(pr); saveRDS(pr, \"/home/s1mi/chr20-qtl2-outputs/batch13_prob.rds\")\n",
      "> \n",
      "> \n",
      "Batch: 14 ... 4.885130250700361 minutes\n",
      "\n",
      "R version 4.3.2 (2023-10-31) -- \"Eye Holes\"\n",
      "Copyright (C) 2023 The R Foundation for Statistical Computing\n",
      "Platform: x86_64-conda-linux-gnu (64-bit)\n",
      "\n",
      "R is free software and comes with ABSOLUTELY NO WARRANTY.\n",
      "You are welcome to redistribute it under certain conditions.\n",
      "Type 'license()' or 'licence()' for distribution details.\n",
      "\n",
      "  Natural language support but running in an English locale\n",
      "\n",
      "R is a collaborative project with many contributors.\n",
      "Type 'contributors()' for more information and\n",
      "'citation()' on how to cite R or R packages in publications.\n",
      "\n",
      "Type 'demo()' for some demos, 'help()' for on-line help, or\n",
      "'help.start()' for an HTML browser interface to help.\n",
      "Type 'q()' to quit R.\n",
      "\n",
      "> library(qtl2); cross <- read_cross2(\"/home/s1mi/tmp-chr20-qtl2-founder-haps/control.yaml\"); pr <- calc_genoprob(cross, error_prob = 0.01, cores = 2); pr <- genoprob_to_alleleprob(pr); saveRDS(pr, \"/home/s1mi/chr20-qtl2-outputs/batch14_prob.rds\")\n",
      "> \n",
      "> \n",
      "Batch: 15 ... 5.367877415949867 minutes\n",
      "\n",
      "R version 4.3.2 (2023-10-31) -- \"Eye Holes\"\n",
      "Copyright (C) 2023 The R Foundation for Statistical Computing\n",
      "Platform: x86_64-conda-linux-gnu (64-bit)\n",
      "\n",
      "R is free software and comes with ABSOLUTELY NO WARRANTY.\n",
      "You are welcome to redistribute it under certain conditions.\n",
      "Type 'license()' or 'licence()' for distribution details.\n",
      "\n",
      "  Natural language support but running in an English locale\n",
      "\n",
      "R is a collaborative project with many contributors.\n",
      "Type 'contributors()' for more information and\n",
      "'citation()' on how to cite R or R packages in publications.\n",
      "\n",
      "Type 'demo()' for some demos, 'help()' for on-line help, or\n",
      "'help.start()' for an HTML browser interface to help.\n",
      "Type 'q()' to quit R.\n",
      "\n",
      "> library(qtl2); cross <- read_cross2(\"/home/s1mi/tmp-chr20-qtl2-founder-haps/control.yaml\"); pr <- calc_genoprob(cross, error_prob = 0.01, cores = 2); pr <- genoprob_to_alleleprob(pr); saveRDS(pr, \"/home/s1mi/chr20-qtl2-outputs/batch15_prob.rds\")\n",
      "> \n",
      "> \n",
      "Batch: 16 ... 5.298072982882149 minutes\n",
      "\n",
      "R version 4.3.2 (2023-10-31) -- \"Eye Holes\"\n",
      "Copyright (C) 2023 The R Foundation for Statistical Computing\n",
      "Platform: x86_64-conda-linux-gnu (64-bit)\n",
      "\n",
      "R is free software and comes with ABSOLUTELY NO WARRANTY.\n",
      "You are welcome to redistribute it under certain conditions.\n",
      "Type 'license()' or 'licence()' for distribution details.\n",
      "\n",
      "  Natural language support but running in an English locale\n",
      "\n",
      "R is a collaborative project with many contributors.\n",
      "Type 'contributors()' for more information and\n",
      "'citation()' on how to cite R or R packages in publications.\n",
      "\n",
      "Type 'demo()' for some demos, 'help()' for on-line help, or\n",
      "'help.start()' for an HTML browser interface to help.\n",
      "Type 'q()' to quit R.\n",
      "\n",
      "> library(qtl2); cross <- read_cross2(\"/home/s1mi/tmp-chr20-qtl2-founder-haps/control.yaml\"); pr <- calc_genoprob(cross, error_prob = 0.01, cores = 2); pr <- genoprob_to_alleleprob(pr); saveRDS(pr, \"/home/s1mi/chr20-qtl2-outputs/batch16_prob.rds\")\n",
      "> \n",
      "> \n",
      "Batch: 17 ... 5.088831246948879 minutes\n",
      "\n",
      "R version 4.3.2 (2023-10-31) -- \"Eye Holes\"\n",
      "Copyright (C) 2023 The R Foundation for Statistical Computing\n",
      "Platform: x86_64-conda-linux-gnu (64-bit)\n",
      "\n",
      "R is free software and comes with ABSOLUTELY NO WARRANTY.\n",
      "You are welcome to redistribute it under certain conditions.\n",
      "Type 'license()' or 'licence()' for distribution details.\n",
      "\n",
      "  Natural language support but running in an English locale\n",
      "\n",
      "R is a collaborative project with many contributors.\n",
      "Type 'contributors()' for more information and\n",
      "'citation()' on how to cite R or R packages in publications.\n",
      "\n",
      "Type 'demo()' for some demos, 'help()' for on-line help, or\n",
      "'help.start()' for an HTML browser interface to help.\n",
      "Type 'q()' to quit R.\n",
      "\n",
      "> library(qtl2); cross <- read_cross2(\"/home/s1mi/tmp-chr20-qtl2-founder-haps/control.yaml\"); pr <- calc_genoprob(cross, error_prob = 0.01, cores = 2); pr <- genoprob_to_alleleprob(pr); saveRDS(pr, \"/home/s1mi/chr20-qtl2-outputs/batch17_prob.rds\")\n",
      "> \n",
      "> \n",
      "Batch: 18 ... 5.172468327915218 minutes\n",
      "\n",
      "R version 4.3.2 (2023-10-31) -- \"Eye Holes\"\n",
      "Copyright (C) 2023 The R Foundation for Statistical Computing\n",
      "Platform: x86_64-conda-linux-gnu (64-bit)\n",
      "\n",
      "R is free software and comes with ABSOLUTELY NO WARRANTY.\n",
      "You are welcome to redistribute it under certain conditions.\n",
      "Type 'license()' or 'licence()' for distribution details.\n",
      "\n",
      "  Natural language support but running in an English locale\n",
      "\n",
      "R is a collaborative project with many contributors.\n",
      "Type 'contributors()' for more information and\n",
      "'citation()' on how to cite R or R packages in publications.\n",
      "\n",
      "Type 'demo()' for some demos, 'help()' for on-line help, or\n",
      "'help.start()' for an HTML browser interface to help.\n",
      "Type 'q()' to quit R.\n",
      "\n",
      "> library(qtl2); cross <- read_cross2(\"/home/s1mi/tmp-chr20-qtl2-founder-haps/control.yaml\"); pr <- calc_genoprob(cross, error_prob = 0.01, cores = 2); pr <- genoprob_to_alleleprob(pr); saveRDS(pr, \"/home/s1mi/chr20-qtl2-outputs/batch18_prob.rds\")\n",
      "> \n",
      "> \n",
      "Batch: 19 ... 5.2715786983006785 minutes\n",
      "\n",
      "R version 4.3.2 (2023-10-31) -- \"Eye Holes\"\n",
      "Copyright (C) 2023 The R Foundation for Statistical Computing\n",
      "Platform: x86_64-conda-linux-gnu (64-bit)\n",
      "\n",
      "R is free software and comes with ABSOLUTELY NO WARRANTY.\n",
      "You are welcome to redistribute it under certain conditions.\n",
      "Type 'license()' or 'licence()' for distribution details.\n",
      "\n",
      "  Natural language support but running in an English locale\n",
      "\n",
      "R is a collaborative project with many contributors.\n",
      "Type 'contributors()' for more information and\n",
      "'citation()' on how to cite R or R packages in publications.\n",
      "\n",
      "Type 'demo()' for some demos, 'help()' for on-line help, or\n",
      "'help.start()' for an HTML browser interface to help.\n",
      "Type 'q()' to quit R.\n",
      "\n",
      "> library(qtl2); cross <- read_cross2(\"/home/s1mi/tmp-chr20-qtl2-founder-haps/control.yaml\"); pr <- calc_genoprob(cross, error_prob = 0.01, cores = 2); pr <- genoprob_to_alleleprob(pr); saveRDS(pr, \"/home/s1mi/chr20-qtl2-outputs/batch19_prob.rds\")\n",
      "> \n",
      "> \n",
      "Batch: 20 ... 4.73374969640087 minutes\n",
      "\n",
      "R version 4.3.2 (2023-10-31) -- \"Eye Holes\"\n",
      "Copyright (C) 2023 The R Foundation for Statistical Computing\n",
      "Platform: x86_64-conda-linux-gnu (64-bit)\n",
      "\n",
      "R is free software and comes with ABSOLUTELY NO WARRANTY.\n",
      "You are welcome to redistribute it under certain conditions.\n",
      "Type 'license()' or 'licence()' for distribution details.\n",
      "\n",
      "  Natural language support but running in an English locale\n",
      "\n",
      "R is a collaborative project with many contributors.\n",
      "Type 'contributors()' for more information and\n",
      "'citation()' on how to cite R or R packages in publications.\n",
      "\n",
      "Type 'demo()' for some demos, 'help()' for on-line help, or\n",
      "'help.start()' for an HTML browser interface to help.\n",
      "Type 'q()' to quit R.\n",
      "\n",
      "> library(qtl2); cross <- read_cross2(\"/home/s1mi/tmp-chr20-qtl2-founder-haps/control.yaml\"); pr <- calc_genoprob(cross, error_prob = 0.01, cores = 2); pr <- genoprob_to_alleleprob(pr); saveRDS(pr, \"/home/s1mi/chr20-qtl2-outputs/batch20_prob.rds\")\n",
      "> \n",
      "> \n",
      "Batch: 21 ... 4.411744385201018 minutes\n",
      "\n",
      "R version 4.3.2 (2023-10-31) -- \"Eye Holes\"\n",
      "Copyright (C) 2023 The R Foundation for Statistical Computing\n",
      "Platform: x86_64-conda-linux-gnu (64-bit)\n",
      "\n",
      "R is free software and comes with ABSOLUTELY NO WARRANTY.\n",
      "You are welcome to redistribute it under certain conditions.\n",
      "Type 'license()' or 'licence()' for distribution details.\n",
      "\n",
      "  Natural language support but running in an English locale\n",
      "\n",
      "R is a collaborative project with many contributors.\n",
      "Type 'contributors()' for more information and\n",
      "'citation()' on how to cite R or R packages in publications.\n",
      "\n",
      "Type 'demo()' for some demos, 'help()' for on-line help, or\n",
      "'help.start()' for an HTML browser interface to help.\n",
      "Type 'q()' to quit R.\n",
      "\n",
      "> library(qtl2); cross <- read_cross2(\"/home/s1mi/tmp-chr20-qtl2-founder-haps/control.yaml\"); pr <- calc_genoprob(cross, error_prob = 0.01, cores = 2); pr <- genoprob_to_alleleprob(pr); saveRDS(pr, \"/home/s1mi/chr20-qtl2-outputs/batch21_prob.rds\")\n",
      "> \n",
      "> \n",
      "Batch: 22 ... 4.215313976900264 minutes\n",
      "\n",
      "R version 4.3.2 (2023-10-31) -- \"Eye Holes\"\n",
      "Copyright (C) 2023 The R Foundation for Statistical Computing\n",
      "Platform: x86_64-conda-linux-gnu (64-bit)\n",
      "\n",
      "R is free software and comes with ABSOLUTELY NO WARRANTY.\n",
      "You are welcome to redistribute it under certain conditions.\n",
      "Type 'license()' or 'licence()' for distribution details.\n",
      "\n",
      "  Natural language support but running in an English locale\n",
      "\n",
      "R is a collaborative project with many contributors.\n",
      "Type 'contributors()' for more information and\n",
      "'citation()' on how to cite R or R packages in publications.\n",
      "\n",
      "Type 'demo()' for some demos, 'help()' for on-line help, or\n",
      "'help.start()' for an HTML browser interface to help.\n",
      "Type 'q()' to quit R.\n",
      "\n",
      "> library(qtl2); cross <- read_cross2(\"/home/s1mi/tmp-chr20-qtl2-founder-haps/control.yaml\"); pr <- calc_genoprob(cross, error_prob = 0.01, cores = 2); pr <- genoprob_to_alleleprob(pr); saveRDS(pr, \"/home/s1mi/chr20-qtl2-outputs/batch22_prob.rds\")\n",
      "> \n",
      "> \n",
      "Batch: 23 ... 4.53186046264988 minutes\n",
      "\n",
      "R version 4.3.2 (2023-10-31) -- \"Eye Holes\"\n",
      "Copyright (C) 2023 The R Foundation for Statistical Computing\n",
      "Platform: x86_64-conda-linux-gnu (64-bit)\n",
      "\n",
      "R is free software and comes with ABSOLUTELY NO WARRANTY.\n",
      "You are welcome to redistribute it under certain conditions.\n",
      "Type 'license()' or 'licence()' for distribution details.\n",
      "\n",
      "  Natural language support but running in an English locale\n",
      "\n",
      "R is a collaborative project with many contributors.\n",
      "Type 'contributors()' for more information and\n",
      "'citation()' on how to cite R or R packages in publications.\n",
      "\n",
      "Type 'demo()' for some demos, 'help()' for on-line help, or\n",
      "'help.start()' for an HTML browser interface to help.\n",
      "Type 'q()' to quit R.\n",
      "\n",
      "> library(qtl2); cross <- read_cross2(\"/home/s1mi/tmp-chr20-qtl2-founder-haps/control.yaml\"); pr <- calc_genoprob(cross, error_prob = 0.01, cores = 2); pr <- genoprob_to_alleleprob(pr); saveRDS(pr, \"/home/s1mi/chr20-qtl2-outputs/batch23_prob.rds\")\n",
      "> \n",
      "> \n",
      "Batch: 24 ... 3.9959222979164526 minutes\n",
      "\n",
      "R version 4.3.2 (2023-10-31) -- \"Eye Holes\"\n",
      "Copyright (C) 2023 The R Foundation for Statistical Computing\n",
      "Platform: x86_64-conda-linux-gnu (64-bit)\n",
      "\n",
      "R is free software and comes with ABSOLUTELY NO WARRANTY.\n",
      "You are welcome to redistribute it under certain conditions.\n",
      "Type 'license()' or 'licence()' for distribution details.\n",
      "\n",
      "  Natural language support but running in an English locale\n",
      "\n",
      "R is a collaborative project with many contributors.\n",
      "Type 'contributors()' for more information and\n",
      "'citation()' on how to cite R or R packages in publications.\n",
      "\n",
      "Type 'demo()' for some demos, 'help()' for on-line help, or\n",
      "'help.start()' for an HTML browser interface to help.\n",
      "Type 'q()' to quit R.\n",
      "\n",
      "> library(qtl2); cross <- read_cross2(\"/home/s1mi/tmp-chr20-qtl2-founder-haps/control.yaml\"); pr <- calc_genoprob(cross, error_prob = 0.01, cores = 2); pr <- genoprob_to_alleleprob(pr); saveRDS(pr, \"/home/s1mi/chr20-qtl2-outputs/batch24_prob.rds\")\n",
      "> \n",
      "> \n",
      "Batch: 25 ... 3.932872584433062 minutes\n",
      "\n",
      "R version 4.3.2 (2023-10-31) -- \"Eye Holes\"\n",
      "Copyright (C) 2023 The R Foundation for Statistical Computing\n",
      "Platform: x86_64-conda-linux-gnu (64-bit)\n",
      "\n",
      "R is free software and comes with ABSOLUTELY NO WARRANTY.\n",
      "You are welcome to redistribute it under certain conditions.\n",
      "Type 'license()' or 'licence()' for distribution details.\n",
      "\n",
      "  Natural language support but running in an English locale\n",
      "\n",
      "R is a collaborative project with many contributors.\n",
      "Type 'contributors()' for more information and\n",
      "'citation()' on how to cite R or R packages in publications.\n",
      "\n",
      "Type 'demo()' for some demos, 'help()' for on-line help, or\n",
      "'help.start()' for an HTML browser interface to help.\n",
      "Type 'q()' to quit R.\n",
      "\n",
      "> library(qtl2); cross <- read_cross2(\"/home/s1mi/tmp-chr20-qtl2-founder-haps/control.yaml\"); pr <- calc_genoprob(cross, error_prob = 0.01, cores = 2); pr <- genoprob_to_alleleprob(pr); saveRDS(pr, \"/home/s1mi/chr20-qtl2-outputs/batch25_prob.rds\")\n",
      "> \n",
      "> \n",
      "Batch: 26 ... 4.306167855666717 minutes\n",
      "\n",
      "R version 4.3.2 (2023-10-31) -- \"Eye Holes\"\n",
      "Copyright (C) 2023 The R Foundation for Statistical Computing\n",
      "Platform: x86_64-conda-linux-gnu (64-bit)\n",
      "\n",
      "R is free software and comes with ABSOLUTELY NO WARRANTY.\n",
      "You are welcome to redistribute it under certain conditions.\n",
      "Type 'license()' or 'licence()' for distribution details.\n",
      "\n",
      "  Natural language support but running in an English locale\n",
      "\n",
      "R is a collaborative project with many contributors.\n",
      "Type 'contributors()' for more information and\n",
      "'citation()' on how to cite R or R packages in publications.\n",
      "\n",
      "Type 'demo()' for some demos, 'help()' for on-line help, or\n",
      "'help.start()' for an HTML browser interface to help.\n",
      "Type 'q()' to quit R.\n",
      "\n",
      "> library(qtl2); cross <- read_cross2(\"/home/s1mi/tmp-chr20-qtl2-founder-haps/control.yaml\"); pr <- calc_genoprob(cross, error_prob = 0.01, cores = 2); pr <- genoprob_to_alleleprob(pr); saveRDS(pr, \"/home/s1mi/chr20-qtl2-outputs/batch26_prob.rds\")\n",
      "> \n",
      "> \n",
      "Batch: 27 ... 4.43262422256715 minutes\n",
      "\n",
      "R version 4.3.2 (2023-10-31) -- \"Eye Holes\"\n",
      "Copyright (C) 2023 The R Foundation for Statistical Computing\n",
      "Platform: x86_64-conda-linux-gnu (64-bit)\n",
      "\n",
      "R is free software and comes with ABSOLUTELY NO WARRANTY.\n",
      "You are welcome to redistribute it under certain conditions.\n",
      "Type 'license()' or 'licence()' for distribution details.\n",
      "\n",
      "  Natural language support but running in an English locale\n",
      "\n",
      "R is a collaborative project with many contributors.\n",
      "Type 'contributors()' for more information and\n",
      "'citation()' on how to cite R or R packages in publications.\n",
      "\n",
      "Type 'demo()' for some demos, 'help()' for on-line help, or\n",
      "'help.start()' for an HTML browser interface to help.\n",
      "Type 'q()' to quit R.\n",
      "\n",
      "> library(qtl2); cross <- read_cross2(\"/home/s1mi/tmp-chr20-qtl2-founder-haps/control.yaml\"); pr <- calc_genoprob(cross, error_prob = 0.01, cores = 2); pr <- genoprob_to_alleleprob(pr); saveRDS(pr, \"/home/s1mi/chr20-qtl2-outputs/batch27_prob.rds\")\n",
      "> \n",
      "> \n",
      "Batch: 28 ... 4.370494974033985 minutes\n",
      "\n",
      "R version 4.3.2 (2023-10-31) -- \"Eye Holes\"\n",
      "Copyright (C) 2023 The R Foundation for Statistical Computing\n",
      "Platform: x86_64-conda-linux-gnu (64-bit)\n",
      "\n",
      "R is free software and comes with ABSOLUTELY NO WARRANTY.\n",
      "You are welcome to redistribute it under certain conditions.\n",
      "Type 'license()' or 'licence()' for distribution details.\n",
      "\n",
      "  Natural language support but running in an English locale\n",
      "\n",
      "R is a collaborative project with many contributors.\n",
      "Type 'contributors()' for more information and\n",
      "'citation()' on how to cite R or R packages in publications.\n",
      "\n",
      "Type 'demo()' for some demos, 'help()' for on-line help, or\n",
      "'help.start()' for an HTML browser interface to help.\n",
      "Type 'q()' to quit R.\n",
      "\n",
      "> library(qtl2); cross <- read_cross2(\"/home/s1mi/tmp-chr20-qtl2-founder-haps/control.yaml\"); pr <- calc_genoprob(cross, error_prob = 0.01, cores = 2); pr <- genoprob_to_alleleprob(pr); saveRDS(pr, \"/home/s1mi/chr20-qtl2-outputs/batch28_prob.rds\")\n",
      "> \n",
      "> \n",
      "Batch: 29 ... 4.498125915382601 minutes\n",
      "\n",
      "R version 4.3.2 (2023-10-31) -- \"Eye Holes\"\n",
      "Copyright (C) 2023 The R Foundation for Statistical Computing\n",
      "Platform: x86_64-conda-linux-gnu (64-bit)\n",
      "\n",
      "R is free software and comes with ABSOLUTELY NO WARRANTY.\n",
      "You are welcome to redistribute it under certain conditions.\n",
      "Type 'license()' or 'licence()' for distribution details.\n",
      "\n",
      "  Natural language support but running in an English locale\n",
      "\n",
      "R is a collaborative project with many contributors.\n",
      "Type 'contributors()' for more information and\n",
      "'citation()' on how to cite R or R packages in publications.\n",
      "\n",
      "Type 'demo()' for some demos, 'help()' for on-line help, or\n",
      "'help.start()' for an HTML browser interface to help.\n",
      "Type 'q()' to quit R.\n",
      "\n",
      "> library(qtl2); cross <- read_cross2(\"/home/s1mi/tmp-chr20-qtl2-founder-haps/control.yaml\"); pr <- calc_genoprob(cross, error_prob = 0.01, cores = 2); pr <- genoprob_to_alleleprob(pr); saveRDS(pr, \"/home/s1mi/chr20-qtl2-outputs/batch29_prob.rds\")\n",
      "> \n",
      "> \n",
      "Batch: 30 ... 5.5711056028997215 minutes\n",
      "\n",
      "R version 4.3.2 (2023-10-31) -- \"Eye Holes\"\n",
      "Copyright (C) 2023 The R Foundation for Statistical Computing\n",
      "Platform: x86_64-conda-linux-gnu (64-bit)\n",
      "\n",
      "R is free software and comes with ABSOLUTELY NO WARRANTY.\n",
      "You are welcome to redistribute it under certain conditions.\n",
      "Type 'license()' or 'licence()' for distribution details.\n",
      "\n",
      "  Natural language support but running in an English locale\n",
      "\n",
      "R is a collaborative project with many contributors.\n",
      "Type 'contributors()' for more information and\n",
      "'citation()' on how to cite R or R packages in publications.\n",
      "\n",
      "Type 'demo()' for some demos, 'help()' for on-line help, or\n",
      "'help.start()' for an HTML browser interface to help.\n",
      "Type 'q()' to quit R.\n",
      "\n",
      "> library(qtl2); cross <- read_cross2(\"/home/s1mi/tmp-chr20-qtl2-founder-haps/control.yaml\"); pr <- calc_genoprob(cross, error_prob = 0.01, cores = 2); pr <- genoprob_to_alleleprob(pr); saveRDS(pr, \"/home/s1mi/chr20-qtl2-outputs/batch30_prob.rds\")\n",
      "> \n",
      "> \n",
      "Batch: 31 ... 5.131772985549954 minutes\n",
      "\n",
      "R version 4.3.2 (2023-10-31) -- \"Eye Holes\"\n",
      "Copyright (C) 2023 The R Foundation for Statistical Computing\n",
      "Platform: x86_64-conda-linux-gnu (64-bit)\n",
      "\n",
      "R is free software and comes with ABSOLUTELY NO WARRANTY.\n",
      "You are welcome to redistribute it under certain conditions.\n",
      "Type 'license()' or 'licence()' for distribution details.\n",
      "\n",
      "  Natural language support but running in an English locale\n",
      "\n",
      "R is a collaborative project with many contributors.\n",
      "Type 'contributors()' for more information and\n",
      "'citation()' on how to cite R or R packages in publications.\n",
      "\n",
      "Type 'demo()' for some demos, 'help()' for on-line help, or\n",
      "'help.start()' for an HTML browser interface to help.\n",
      "Type 'q()' to quit R.\n",
      "\n",
      "> library(qtl2); cross <- read_cross2(\"/home/s1mi/tmp-chr20-qtl2-founder-haps/control.yaml\"); pr <- calc_genoprob(cross, error_prob = 0.01, cores = 2); pr <- genoprob_to_alleleprob(pr); saveRDS(pr, \"/home/s1mi/chr20-qtl2-outputs/batch31_prob.rds\")\n",
      "> \n",
      "> \n",
      "Batch: 32 ... 5.087727682616484 minutes\n",
      "\n",
      "R version 4.3.2 (2023-10-31) -- \"Eye Holes\"\n",
      "Copyright (C) 2023 The R Foundation for Statistical Computing\n",
      "Platform: x86_64-conda-linux-gnu (64-bit)\n",
      "\n",
      "R is free software and comes with ABSOLUTELY NO WARRANTY.\n",
      "You are welcome to redistribute it under certain conditions.\n",
      "Type 'license()' or 'licence()' for distribution details.\n",
      "\n",
      "  Natural language support but running in an English locale\n",
      "\n",
      "R is a collaborative project with many contributors.\n",
      "Type 'contributors()' for more information and\n",
      "'citation()' on how to cite R or R packages in publications.\n",
      "\n",
      "Type 'demo()' for some demos, 'help()' for on-line help, or\n",
      "'help.start()' for an HTML browser interface to help.\n",
      "Type 'q()' to quit R.\n",
      "\n",
      "> library(qtl2); cross <- read_cross2(\"/home/s1mi/tmp-chr20-qtl2-founder-haps/control.yaml\"); pr <- calc_genoprob(cross, error_prob = 0.01, cores = 2); pr <- genoprob_to_alleleprob(pr); saveRDS(pr, \"/home/s1mi/chr20-qtl2-outputs/batch32_prob.rds\")\n"
     ]
    }
   ],
   "source": [
    "for i in range(n_batches):\n",
    "    tic = time.perf_counter()\n",
    "    geno_df = geno.iloc[:, i:i+batch_size]\n",
    "    samples = geno_df.columns.to_list()\n",
    "    covar_df = pd.DataFrame({'id': samples, 'generations': [90] *  len(samples)})\n",
    "    geno_df.to_csv(f'{tmp_dir}/geno.csv', index=True)\n",
    "    covar_df.to_csv(f'{tmp_dir}/covar.csv', index=False)\n",
    "    cmd = qtl_command(f'{output_dir}/batch{i}_prob.rds', n_cores = 2)\n",
    "    subprocess.run(f\"R -e '{cmd}'\", shell=True)\n",
    "    toc = time.perf_counter()\n",
    "    print(\"Batch:\", i+1, \"...\", (toc-tic)/60, \"minutes\")"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.12.0"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
