---
title: "Home Page for Enformer Human and Rat Cross-species Analysis"
descriptions: Updates, links to analysis in the rat project
format: html
---

## Updates this Week

* [Found human head output (7,611 tracks)](https://sabrina-dl.hakyimlab.org/posts/home/s1mi/Github/deep-learning-in-genomics/posts/2023-09-25-baskerville-tutorial/)
* Started running predictions on human reference genome for 981 genes
    * No results yet, waiting in the queue
    * Average GPU time per prediction doubled? (~2.5 -> ~5 seconds)
* [Set up personalized prediction pipeline](https://sabrina-dl.hakyimlab.org/posts/2023-09-26-borzoi-personalized-test/geuvadis_personalized_test)

## Introduction

### Data Inventory

* Genotype and gene expression data for 340 rats in the Br study
    * Mean gene expression used as ground truth reference gene expression
* GTEx gene expression data used as ground truth reference gene expression
* Rat reference genome as input to Enformer to predict reference gene expression 
* Human reference genome as input to Enformer to predict reference gene expression

### Goals

* Determine whether Enformer models (human and mouse) can predict expression across genes
* Use genotypes to predict gene expression in the Br rats with the Enformer models


## Correlation across genes

We are currently using the CAGE:Brain human track at a gene's TSS to estimate gene expression in rats. For all GTEx gene expression comparisons, we used mean TPM in brain cortex tissue. We want to make 3 types of comparison:

1. $\rho^h_{Hh}$: Correlation between human reference gene expression predicted with Enformer human head and GTEx mean gene expression
2. $\rho^h_{Rr}$: Correlation between rat reference gene expression predicted with Enformer human head and Br reference gene expression
3. $\rho^h_{HR}$: Correlation between rat reference gene expression and human reference gene expression

Enformer is optimized to predict variation across the genome, so we use $\rho^h_{Hh}$ as the baseline performance. Intuitively, we hope to see $\rho^h_{Rr} = \rho^h_{HR} \rho^h_{Hh}$.

[Personal transcriptome variation is poorly explained by current genomic deep learning models, Huang et al](https://www.biorxiv.org/content/10.1101/2023.06.30.547100v1.full.pdf) performed the same analysis on GEUVADIS LCL gene expression data and found a correlation of 0.57

### Analysis Summary


1. [$\rho^h_{Hh} = 0.539$](https://sabrina-dl.hakyimlab.org/posts/2023-08-22-correlation-between-enformer-and-gtex-expression-across-genes-by-chromosome/)
2. [$\rho^h_{Rr} = 0.453$](https://sabrina-dl.hakyimlab.org/posts/2023-08-22-correlation-across-rn7-genes-by-chromosome/)
3. [$\rho^h_{HR} = 0.886$](https://sabrina-dl.hakyimlab.org/posts/2023-09-01-comparing-human-vs-rat-mean-expression-across-genes)

Other analysis includes: 
1. Validating Huang et al.'s [correlation of 0.57](https://sabrina-dl.hakyimlab.org/posts/2023-09-06-validating-Huang-et-al) in human LCL tissues, I even obtained a higher correlation of 0.77 when I included all GEUVADIS genes

## Correlation across individuals

I added support for unphased VCFs, which cuts the number of Enformer runs in half by averaging the two haplotypes. I also added options to output mouse head predictions or both.

* [Results for 868 genes selected for significant eQTLs and observed variation](https://sabrina-dl.hakyimlab.org/posts/2023-08-31-Br-personalized-prediction-on-more-genes/)


