{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "79026aef",
   "metadata": {},
   "source": [
    "---\n",
    "title: Computing Haplotype for Probabilities 8K HS Rats\n",
    "author: Sabrina Mi\n",
    "date: 7/31/2025\n",
    "---\n",
    "# Preparing 8K rat genotypes with body lenght, BMI data\n",
    "\n",
    "## Write samples file\n",
    "\n",
    "First, we obtain a list of samples with body length and BMI data, then intersect with genotypes rats."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "1430c559",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/tmp/ipykernel_3687900/3302687599.py:3: DtypeWarning: Columns (0) have mixed types. Specify dtype option on import or set low_memory=False.\n",
      "  all_pheno = pd.read_csv(\"/home/s1mi/enformer_rat_data/phenotypes/ALLTRAITSALLNORMALIZES_19jul24.csv\",\n"
     ]
    }
   ],
   "source": [
    "import pandas as pd\n",
    "all_pheno = pd.read_csv(\"/home/s1mi/enformer_rat_data/phenotypes/ALLTRAITSALLNORMALIZES_19jul24.csv\", \n",
    "                        usecols = ['rfid', 'dissection:regressedlr_length_w_tail_cm', 'dissection:regressedlr_bmi_w_tail'],\n",
    "                        index_col = 'rfid')\n",
    "all_pheno.columns = ['bodylen', 'bmi']\n",
    "samples = list(all_pheno[all_pheno['bodylen'].notna()].index)\n",
    "# with open(\"bodylen_bmisamples.txt\", \"w\") as f:\n",
    "#     f.write(\"\\n\".join(pheno_rats))"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "aa0c82fc",
   "metadata": {},
   "source": [
    "\n",
    "I downloaded the [bed, bim, fam files](https://library.ucsd.edu/dc/object/bb5610743d), then make basic modifications to the files.\n",
    "\n",
    "## Filter Samples\n",
    "\n",
    "Filter to the 8K rats with phenotype data using `plink`.\n",
    "\n",
    "```\n",
    "plink2 --bfile bb5610743d --keep samples.txt --export vcf bgz --out bodylen_bmi_8K_samples\n",
    "```\n",
    "\n",
    "`plink2` automatically generates new IDs of the format `FID_IID`, so we first need to remove the duplicate ID in the VCF header.\n",
    "\n",
    "```\n",
    "bcftools query -l bodylen_bmi_8K_samples.vcf.gz > samples_FID_IID.txt\n",
    "awk -F'_' '{print $0 \"\\t\" $1}' samples_FID_IID.txt > rename_samples.txt\n",
    "bcftools reheader -s rename_samples.txt -o renamed.vcf.gz bodylen_bmi_8K_samples.vcf.gz\n",
    "```\n",
    "\n",
    "```\n",
    "bcftools view renamed.vcf.gz | \\\n",
    "awk 'BEGIN{FS=OFS=\"\\t\"} /^#/ {print; next} {$1=\"chr\"$1; $3=\"chr\"$3; print}' | \\\n",
    "bgzip > bodylen_bmi.vcf.gz\n",
    "bcftools index -t bodylen_bmi.vcf.gz\n",
    "rm renamed.vcf.gz\n",
    "```\n",
    "\n",
    "Some of the samples with phenotype data are not in the genotype file, so `plink2` keeps only the overlapping samples. We generate a list of current samples.\n",
    "\n",
    "```\n",
    "bcftools query -l bodylen_bmi_8K_samples.vcf.gz > /home/s1mi/Github/deep-learning-in-genomics/posts/2025-07-22-bmi-bodylen-rats-haplotype-probabilities/samples.txt\n",
    "```\n",
    "\n",
    "## Split VCF by Chromosome\n",
    "```\n",
    "mkdir ~/enformer_rat_data/genotypes/bodylen_bmi_VCFs\n",
    "# Split VCF by chromosome\n",
    "vcf_in=bodylen_bmi.vcf.gz\n",
    "\n",
    "vcf_out_prefix=~/enformer_rat_data/genotypes/bodylen_bmi_VCFs/chr\n",
    "\n",
    "for i in {1..20}\n",
    "do\n",
    "    echo \"Working on chromosome ${i}...\"\n",
    "    bcftools view ${vcf_in} --regions chr${i} -o ${vcf_out_prefix}${i}.vcf.gz -Oz\n",
    "done\n",
    "```\n",
    "\n",
    "## Index VCFs\n",
    "\n",
    "```\n",
    "for i in {1..20}\n",
    "do\n",
    "    echo \"Indexing chromosome ${i}...\"\n",
    "    bcftools index -t ${vcf_out_prefix}${i}.vcf.gz\n",
    "done\n",
    "```\n",
    "\n",
    "# Compute Haplotype Probabilities\n",
    "\n",
    "Because the genotype files are so large, we need to split them into batches of 15 individuals in order to run qtl2."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "e761c03c",
   "metadata": {},
   "outputs": [],
   "source": [
    "import os\n",
    "with open(\"samples.txt\", \"r\") as f:\n",
    "    samples = f.read().splitlines()\n",
    "batch_size = 20\n",
    "n_batches = len(samples)//batch_size\n",
    "data_dir = '/eagle/AIHPC4Edu/sabrina/scratch/bodylen_bmi_VCFs_for_qtl2'\n",
    "with open(f'{data_dir}/batches.txt', \"w\") as f:\n",
    "    for i in range(n_batches):\n",
    "        batch_str = \",\".join(samples[batch_size*i:batch_size*(i+1)])\n",
    "        f.write(f'{batch_str}\\t-\\tbatch{i}\\n')\n",
    "        "
   ]
  },
  {
   "cell_type": "markdown",
   "id": "d54af00b",
   "metadata": {},
   "source": [
    "## Split VCF by batch\n",
    "\n",
    "```\n",
    "module use /soft/modulefiles; module load conda; conda activate genomics\n",
    "output_dir=\"/eagle/AIHPC4Edu/sabrina/scratch/bodylen_bmi_VCFs_for_qtl2\"\n",
    "for i in {1..20}; do mkdir -p ${output_dir}/chr${i}_by_batch; done\n",
    "log_file=\"split_VCF_by_batch.log\"\n",
    "for i in {1..20}\n",
    "do\n",
    "    start_time=$(date +%s)\n",
    "    echo \"Working on chromosome ${i}...\" >> \"$log_file\"\n",
    "    vcf_in=~/enformer_rat_data/genotypes/bodylen_bmi_VCFs/chr${i}.vcf.gz\n",
    "\n",
    "    bcftools +split ${vcf_in} -S ${output_dir}/batches.txt --output-type z --output ${output_dir}/chr${i}_by_batch\n",
    "\n",
    "    end_time=$(date +%s)\n",
    "    duration=$((end_time - start_time))\n",
    "    echo \"Completed chromosome ${i} in ${duration}s\" >> \"$log_file\"\n",
    "done\n",
    "```\n",
    "\n",
    "## Run qtl2\n",
    "\n",
    "This is example code for batch 0 (all chromosomes), that I ran in an interactive job.\n",
    "\n",
    "```\n",
    "j=0\n",
    "\n",
    "module use /soft/modulefiles; module load conda\n",
    "data_dir=/home/s1mi/qtl2_data\n",
    "scratch=/eagle/AIHPC4Edu/sabrina/scratch\n",
    "samples_dir=${scratch}/bodylen_bmi_VCFs_for_qtl2\n",
    "founders_dir=/home/s1mi/enformer_rat_data/genotypes/FounderVCFs\n",
    "output_dir=${scratch}/qtl2_outputs\n",
    "code_dir=/home/s1mi/Github/deep-learning-in-genomics/posts/2025-07-22-bmi-bodylen-rats-haplotype-probabilities\n",
    "\n",
    "for i in {16..20}\n",
    "do\n",
    "    echo \"Working on chromosome ${i}...\"\n",
    "    start_time=$(date +%s)\n",
    "    conda run -n ml-python python ${code_dir}/make_qtl2_inputs.py ${samples_dir}/chr${i}_by_batch/batch${j}.vcf.gz ${founders_dir}/chr${i}.vcf.gz ${output_dir}/chr${i}_by_batch/batch${j}_probs.rds --working-dir ${scratch}/chr${i}_qtl2_inputs/batch${j} --gmap-dir ${data_dir}/genetic_map --cores 32\n",
    "\n",
    "    end_time=$(date +%s)\n",
    "    duration=$((end_time - start_time))\n",
    "    echo \"Prepared chromosome ${i} inputs for qtl2 in ${duration}s\"\n",
    "\n",
    "    cd ${scratch}/chr${i}_qtl2_founder_haps\n",
    "    conda run -n genomics Rscript ${scratch}/chr${i}_qtl2_founder_haps/qtl2_calculate_prob.R\n",
    "\n",
    "    end_time=$(date +%s)\n",
    "    duration=$((end_time - start_time))\n",
    "    echo \"Computed chromosome ${i} haplotype probabilities in ${duration}s\"\n",
    "done\n",
    "\n",
    "```\n",
    "\n",
    "I ran this in two interactive job, 2 hours total, sequentially. Since the writing input files take much less time, I put it in it's own PBS script, `prepare_qtl2_inputs.pbs`, that runs the Python code in parallel by chromosome and batch. Then `run_qtl2.pbs` executes the R code in parallel.\n",
    "\n",
    "First create a file of all chromosome, batch pairs.\n",
    "\n",
    "```\n",
    "job_list=${PBS_O_WORKDIR}/job_list.txt\n",
    "for i in {1..20}; do\n",
    "  for j in {1..434}; do\n",
    "    echo $i $j >> $job_list\n",
    "  done\n",
    "done\n",
    "```\n",
    "`qsub prepare_qtl2_inputs.pbs`\n",
    "\n",
    "\n",
    "`qsub -I -A AIHPC4EDU -l walltime=1:00:00 -l filesystems=home:eagle -q debug`\n"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "ml-python",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.10.11"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
