---
title: Borzoi vs Enformer Comparisons
---

## Genome-wide Performance

We feel confident enough in Borzoi's performance across the reference genome, both in CAGE:Brain and RNA-Seq:Brain tracks, to say it performs as well, if not better, than Enformer.

[Test on ~1000 Genes](https://sabrina-dl.hakyimlab.org/posts/2023-09-25-borzoi-cage-test/)


## Cross-Population Performance

I double-checked the borzoi code and tested various adjustments thoroughly enough to rule out that the low correlations are from errors on my end. For each of the gene tests, I used GEUVADIS genotypes (N=455) and LCL gene expression data.


[First test on 3 genes](https://sabrina-dl.hakyimlab.org/posts/2023-10-04-personalized-test-on-a-few-genes/)

* I used *human* GEUVADIS data to compare predicted and observed expression. Based off my (growing) understanding of the Borzoi tracks, I decided tracks 870 and 871 (CAGE:B lymphoblastoid cell line, + and - stranded) for CAGE predictions and track 7531 (RNA-Seq:Blood, GTEX-1I4MK-0002-SM-EZ6M9) 
* All 3 genes had very low correlations <0.1 (CAGE and RNA-Seq), but I noticed RNA-Seq had higher correlation than CAGE for all genes

[Focusing on SLFN5](https://sabrina-dl.hakyimlab.org/posts/2023-10-18-borzoi-personalized-on-Huang-et-al-high-performance-gene/), Enformer performed well for this gene in Huang et al's paper

* Best correlation so far: Spearman R=0.160 using only + strand

[Compare to Enformer](https://sabrina-dl.hakyimlab.org/posts/2023-10-20-borzoi-vs-enformer-personalized-prediction)

* Borzoi's CAGE tracks for LCL are worse than Enformer's at predicting expression

![borzoi vs enformer on GEUVADIS LCL tissue](borzoi_enformer_scatter.png)
