{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "---\n",
    "title: \"Running Enformer across all rn7 genes\"\n",
    "author: \"Sabrina Mi\"\n",
    "description: \"Our pipeline inputs to run Enformer on the reference genome and sampled rat genotypes\"\n",
    "date: \"8/10/23\"\n",
    "---\n",
    "\n",
    "## Collect Gene Intervals"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "import random"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/tmp/ipykernel_206323/2449699668.py:2: DtypeWarning: Columns (2) have mixed types. Specify dtype option on import or set low_memory=False.\n",
      "  annot_df = pd.read_csv('/home/s1mi/enformer_rat_data/rn7.gene.txt', sep ='\\t',  index_col='geneId')\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>geneSymbol</th>\n",
       "      <th>chromosome</th>\n",
       "      <th>start</th>\n",
       "      <th>end</th>\n",
       "      <th>strand</th>\n",
       "      <th>tss</th>\n",
       "      <th>description</th>\n",
       "      <th>hasEqtl</th>\n",
       "      <th>expr_BLA</th>\n",
       "      <th>expr_Brain</th>\n",
       "      <th>...</th>\n",
       "      <th>sqtl_BLA</th>\n",
       "      <th>sqtl_Brain</th>\n",
       "      <th>sqtl_Eye</th>\n",
       "      <th>sqtl_IL</th>\n",
       "      <th>sqtl_LHb</th>\n",
       "      <th>sqtl_NAcc</th>\n",
       "      <th>sqtl_NAcc2</th>\n",
       "      <th>sqtl_OFC</th>\n",
       "      <th>sqtl_PL</th>\n",
       "      <th>sqtl_PL2</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>geneId</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>ENSRNOG00000066169</th>\n",
       "      <td>ENSRNOG00000066169</td>\n",
       "      <td>1</td>\n",
       "      <td>36112690</td>\n",
       "      <td>36122387</td>\n",
       "      <td>-</td>\n",
       "      <td>36122387</td>\n",
       "      <td>NaN</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>...</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSRNOG00000070168</th>\n",
       "      <td>Olr56</td>\n",
       "      <td>1</td>\n",
       "      <td>157231467</td>\n",
       "      <td>157232417</td>\n",
       "      <td>+</td>\n",
       "      <td>157231467</td>\n",
       "      <td>olfactory receptor family 51 subfamily F membe...</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>...</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSRNOG00000070901</th>\n",
       "      <td>Irgq</td>\n",
       "      <td>1</td>\n",
       "      <td>80123925</td>\n",
       "      <td>80131881</td>\n",
       "      <td>+</td>\n",
       "      <td>80123925</td>\n",
       "      <td>immunity-related GTPase Q</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>...</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>3 rows Ã— 68 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                            geneSymbol chromosome      start        end  \\\n",
       "geneId                                                                    \n",
       "ENSRNOG00000066169  ENSRNOG00000066169          1   36112690   36122387   \n",
       "ENSRNOG00000070168               Olr56          1  157231467  157232417   \n",
       "ENSRNOG00000070901                Irgq          1   80123925   80131881   \n",
       "\n",
       "                   strand        tss  \\\n",
       "geneId                                 \n",
       "ENSRNOG00000066169      -   36122387   \n",
       "ENSRNOG00000070168      +  157231467   \n",
       "ENSRNOG00000070901      +   80123925   \n",
       "\n",
       "                                                          description  \\\n",
       "geneId                                                                  \n",
       "ENSRNOG00000066169                                                NaN   \n",
       "ENSRNOG00000070168  olfactory receptor family 51 subfamily F membe...   \n",
       "ENSRNOG00000070901                          immunity-related GTPase Q   \n",
       "\n",
       "                    hasEqtl  expr_BLA  expr_Brain  ...  sqtl_BLA  sqtl_Brain  \\\n",
       "geneId                                             ...                         \n",
       "ENSRNOG00000066169     True      True        True  ...     False       False   \n",
       "ENSRNOG00000070168     True     False        True  ...     False       False   \n",
       "ENSRNOG00000070901     True      True        True  ...     False       False   \n",
       "\n",
       "                    sqtl_Eye  sqtl_IL  sqtl_LHb  sqtl_NAcc  sqtl_NAcc2  \\\n",
       "geneId                                                                   \n",
       "ENSRNOG00000066169     False    False     False      False       False   \n",
       "ENSRNOG00000070168     False    False     False      False       False   \n",
       "ENSRNOG00000070901     False    False     False      False       False   \n",
       "\n",
       "                    sqtl_OFC  sqtl_PL  sqtl_PL2  \n",
       "geneId                                           \n",
       "ENSRNOG00000066169     False    False     False  \n",
       "ENSRNOG00000070168     False    False     False  \n",
       "ENSRNOG00000070901     False    False     False  \n",
       "\n",
       "[3 rows x 68 columns]"
      ]
     },
     "execution_count": 3,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# gene annotation\n",
    "annot_df = pd.read_csv('/home/s1mi/enformer_rat_data/rn7.gene.txt', sep ='\\t',  index_col='geneId')\n",
    "annot_df.head(3)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/tmp/ipykernel_206323/3189498692.py:1: DtypeWarning: Columns (0) have mixed types. Specify dtype option on import or set low_memory=False.\n",
      "  obs_gene_expr = pd.read_csv(\"/home/s1mi/enformer_rat_data/Brain.rn7.expr.tpm.bed\", sep=\"\\t\", header=0, index_col='gene_id')\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>#chr</th>\n",
       "      <th>start</th>\n",
       "      <th>end</th>\n",
       "      <th>000789972A</th>\n",
       "      <th>0007899884</th>\n",
       "      <th>000789FF64</th>\n",
       "      <th>00079298B1</th>\n",
       "      <th>00079299B5</th>\n",
       "      <th>0007929A79</th>\n",
       "      <th>0007929B36</th>\n",
       "      <th>...</th>\n",
       "      <th>0007899761</th>\n",
       "      <th>0007899866</th>\n",
       "      <th>00078A00E6</th>\n",
       "      <th>0007929938</th>\n",
       "      <th>0007929956</th>\n",
       "      <th>00079FF7D6</th>\n",
       "      <th>0007899826</th>\n",
       "      <th>0007929917</th>\n",
       "      <th>00077E82E3</th>\n",
       "      <th>00078A18A7</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>gene_id</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>ENSRNOG00000065394</th>\n",
       "      <td>1</td>\n",
       "      <td>31563</td>\n",
       "      <td>31564</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSRNOG00000065356</th>\n",
       "      <td>1</td>\n",
       "      <td>41634</td>\n",
       "      <td>41635</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSRNOG00000070568</th>\n",
       "      <td>1</td>\n",
       "      <td>76833</td>\n",
       "      <td>76834</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.01</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSRNOG00000065706</th>\n",
       "      <td>1</td>\n",
       "      <td>184235</td>\n",
       "      <td>184236</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.18</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSRNOG00000069644</th>\n",
       "      <td>1</td>\n",
       "      <td>272252</td>\n",
       "      <td>272253</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows Ã— 343 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                   #chr   start     end  000789972A  0007899884  000789FF64  \\\n",
       "gene_id                                                                       \n",
       "ENSRNOG00000065394    1   31563   31564         0.0         0.0         0.0   \n",
       "ENSRNOG00000065356    1   41634   41635         0.0         0.0         0.0   \n",
       "ENSRNOG00000070568    1   76833   76834         0.0         0.0         0.0   \n",
       "ENSRNOG00000065706    1  184235  184236         0.0         0.0         0.0   \n",
       "ENSRNOG00000069644    1  272252  272253         0.0         0.0         0.0   \n",
       "\n",
       "                    00079298B1  00079299B5  0007929A79  0007929B36  ...  \\\n",
       "gene_id                                                             ...   \n",
       "ENSRNOG00000065394         0.0        0.00         0.0         0.0  ...   \n",
       "ENSRNOG00000065356         0.0        0.00         0.0         0.0  ...   \n",
       "ENSRNOG00000070568         0.0        0.00         0.0         0.0  ...   \n",
       "ENSRNOG00000065706         0.0        0.18         0.0         0.0  ...   \n",
       "ENSRNOG00000069644         0.0        0.00         0.0         0.0  ...   \n",
       "\n",
       "                    0007899761  0007899866  00078A00E6  0007929938  \\\n",
       "gene_id                                                              \n",
       "ENSRNOG00000065394         0.0         0.0         0.0         0.0   \n",
       "ENSRNOG00000065356         0.0         0.0         0.0         0.0   \n",
       "ENSRNOG00000070568         0.0         0.0         0.0         0.0   \n",
       "ENSRNOG00000065706         0.0         0.0         0.0         0.0   \n",
       "ENSRNOG00000069644         0.0         0.0         0.0         0.0   \n",
       "\n",
       "                    0007929956  00079FF7D6  0007899826  0007929917  \\\n",
       "gene_id                                                              \n",
       "ENSRNOG00000065394         0.0        0.00         0.0         0.0   \n",
       "ENSRNOG00000065356         0.0        0.00         0.0         0.0   \n",
       "ENSRNOG00000070568         0.0        0.01         0.0         0.0   \n",
       "ENSRNOG00000065706         0.0        0.00         0.0         0.0   \n",
       "ENSRNOG00000069644         0.0        0.00         0.0         0.0   \n",
       "\n",
       "                    00077E82E3  00078A18A7  \n",
       "gene_id                                     \n",
       "ENSRNOG00000065394         0.0         0.0  \n",
       "ENSRNOG00000065356         0.0         0.0  \n",
       "ENSRNOG00000070568         0.0         0.0  \n",
       "ENSRNOG00000065706         0.0         0.0  \n",
       "ENSRNOG00000069644         0.0         0.0  \n",
       "\n",
       "[5 rows x 343 columns]"
      ]
     },
     "execution_count": 4,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "obs_gene_expr = pd.read_csv(\"/home/s1mi/enformer_rat_data/Brain.rn7.expr.tpm.bed\", sep=\"\\t\", header=0, index_col='gene_id')\n",
    "obs_gene_expr.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "array([1, 10, 11, 12, 13, 14, 15, 16, 17, 18, 19, 2, 20, 3, 4, 5, 6, 7, 8,\n",
       "       9, '9', 'JACYVU010000238.1', 'JACYVU010000315.1',\n",
       "       'JACYVU010000319.1', 'JACYVU010000493.1', 'JACYVU010000587.1',\n",
       "       'JACYVU010000589.1', 'JACYVU010000619.1', 'JACYVU010000634.1',\n",
       "       'JACYVU010000642.1', 'JACYVU010000653.1', 'JACYVU010000665.1',\n",
       "       'JACYVU010000706.1', 'JACYVU010000731.1', 'JACYVU010000732.1',\n",
       "       'JACYVU010000738.1', 'JACYVU010000744.1', 'JACYVU010000754.1',\n",
       "       'MT', 'MU150189.1', 'MU150193.1', 'MU150196.1', 'MU150200.1',\n",
       "       'MU150203.1', 'MU150220.1', 'MU150222.1', 'MU150223.1', 'X', 'Y'],\n",
       "      dtype=object)"
      ]
     },
     "execution_count": 7,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "obs_gene_expr['#chr'].unique()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "We want to filter to genes on chromosome 1 through 20 and X and Y."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 26,
   "metadata": {},
   "outputs": [],
   "source": [
    "keys = [\"1\", \"2\", \"3\", \"4\", \"5\", \"6\", \"7\", \"8\", \"9\", \"10\", \"11\", \"12\", \"13\", \"14\", \"15\", \"16\", \"17\", \"18\", \"19\", \"20\", \"X\", \"Y\"]\n",
    "\n",
    "with open(\"metadata/intervals.txt\", \"a\") as f:\n",
    "    for gene in obs_gene_expr.index:\n",
    "        gene_annot = annot_df.loc[gene]\n",
    "        chromosome = str(gene_annot['chromosome'])\n",
    "        if chromosome in keys:\n",
    "            tss = gene_annot['tss']\n",
    "            interval = f\"chr{chromosome}_{tss}_{tss}\"\n",
    "            f.write(interval + '\\n')\n",
    "            \n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Sample Individuals\n",
    "\n",
    "We \"randomly\" sampled three rats until we had a set that had at least some variety in color and sex."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>coatcolor</th>\n",
       "      <th>tissues</th>\n",
       "      <th>sex</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>rat_id</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>00079FE6B4</th>\n",
       "      <td>BLACKHOOD</td>\n",
       "      <td>Brain</td>\n",
       "      <td>F</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>0007899981</th>\n",
       "      <td>BROWNHOOD</td>\n",
       "      <td>Brain</td>\n",
       "      <td>M</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>00078A19B2</th>\n",
       "      <td>BROWN</td>\n",
       "      <td>Brain</td>\n",
       "      <td>M</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "            coatcolor tissues sex\n",
       "rat_id                           \n",
       "00079FE6B4  BLACKHOOD   Brain   F\n",
       "0007899981  BROWNHOOD   Brain   M\n",
       "00078A19B2      BROWN   Brain   M"
      ]
     },
     "execution_count": 31,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "rat_info = pd.read_csv(\"/home/s1mi/enformer_rat_data/rn7.RatGTEx_rats.tsv\", sep = \"\\t\", index_col = \"rat_id\")\n",
    "all_individuals = obs_gene_expr.columns.tolist()[3:]\n",
    "sample = random.sample(all_individuals, 3)\n",
    "rat_info.loc[sample]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "with open(\"metadata/individuals.txt\", \"a\") as f:\n",
    "    for individual in sample:\n",
    "        f.write(individual + '\\n')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Run Enformer Pipeline\n",
    "\n",
    "\n",
    "Run personalized for sampled rats:\n",
    "```\n",
    "module load conda\n",
    "\n",
    "conda activate /lus/grand/projects/TFXcan/imlab/shared/software/conda_envs/enformer-predict-tools\n",
    "\n",
    "cd /home/s1mi/Github/shared_pipelines/enformer_pipeline\n",
    "\n",
    "python3 scripts/enformer_predict.py --parameters /home/s1mi/Github/deep-learning-in-genomics/posts/2023-08-10-correlation-across-genes/all_genes_personalized_config.json\n",
    "\n",
    "```\n",
    "\n",
    "Run on reference genome:\n",
    "\n",
    "```\n",
    "module load conda\n",
    "\n",
    "conda activate /lus/grand/projects/TFXcan/imlab/shared/software/conda_envs/enformer-predict-tools\n",
    "\n",
    "cd /home/s1mi/Github/shared_pipelines/enformer_pipeline\n",
    "\n",
    "python3 scripts/enformer_predict.py --parameters /home/s1mi/Github/deep-learning-in-genomics/posts/2023-08-10-correlation-across-genes/all_genes_reference_config.json\n",
    "```"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "base",
   "language": "python",
   "name": "base"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.8.13"
  },
  "orig_nbformat": 4
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
