---
title: RatXcan run on 324 Br Rats
author: Sabrina Mi
date: 08/17/2025
---

## Preliminary Computations

### Predict Expression

```
PRE="/Users/sabrinami/Desktop/RatXcan"
OUTPUT="$PRE/output"
MODEL="$PRE/models"
GENO_PREFIX="$PRE/Br_data/genotype/Br_bodylen_bmi"
METAXCAN="/Users/sabrinami/Github/MetaXcan"
```

```
conda activate imlabtools
tissues="AC IL LH PL VO"
for model in $tissues; do
  python ${METAXCAN}/software/Predict.py \
  --model_db_path ${MODEL}/${model}-filtered.db \
  --model_db_snp_key rsid \
  --vcf_genotypes ${GENO_PREFIX}.vcf.gz \
  --vcf_mode genotyped \
  --liftover ${PRE}/rn7ToRn6.over.chain.gz \
  --on_the_fly_mapping METADATA "{}_{}_{}_{}" \
  --prediction_output $OUTPUT/Br_bodylen_bmi_${model}__predict.txt  \
  --prediction_summary_output $OUTPUT/Br_bodylen_bmi_${model}__summary.txt \
  --throw
done

```

### Compute Genetic Relatedness and Heritability

```
## Compute GRM
gcta --bfile $GENO_PREFIX --make-grm-bin --out $OUTPUT/Br_bodylen_bmi
## Compute h2
## argument --mpheno specifies fam column to be used as phenotype, 1 for bodylength and 2 for bmi
gcta --grm $OUTPUT/Br_bodylen_bmi --reml --pheno $PRE/Br_data/phenotype/pheno.fam --mpheno 1 --out $OUTPUT/bodylen_h2
```

## Prepare Inputs for Mixed Effects Model

```{r libraries}
## compare observed correlation with null correlation plotting functions
suppressMessages(devtools::source_gist("a925fea01b365a8c605e")) ## qqR fn https://gist.github.com/hakyim/a925fea01b365a8c605e
suppressMessages(devtools::source_gist("38431b74c6c0bf90c12f")) ## qqunif https://gist.github.com/hakyim/38431b74c6c0bf90c12f
## ratxcan functions
suppressMessages(devtools::source_gist("115403f16bec0a0e871f3616d552ce9b")) ## https://gist.github.com/hakyim/115403f16bec0a0e871f3616d552ce9b 

suppressMessages(library(tidyverse))
suppressMessages(library(glue))
suppressMessages(library(readr))
suppressMessages(library(biomaRt))

## to install qvalue
## if (!require("BiocManager", quietly = TRUE))
   ## install.packages("BiocManager")
## BiocManager::install("biomaRt")

```


```{r paths}
PRE <- "~/Desktop/RatXcan"
OUTPUT <- glue("{PRE}/output")
```

### Read in GRM, h2, and phenotypes


```{r read gcta outputs}
## read grm
grm_mat <- read_GRMBin(glue("{OUTPUT}/Br_bodylen_bmi.grm"))
grm_id <- read_tsv(glue("{OUTPUT}/Br_bodylen_bmi.grm.id"), col_names = FALSE)
names(grm_id) <- c("FID","IID")
## read h2
tempo <- read_tsv(glue("{OUTPUT}/bodylen_h2.hsq")) %>% filter(Source=="V(G)/Vp")
bodylen_h2 <- tempo %>% pull(Variance)
bodylen_se <- tempo %>% pull(SE)
## read phenotype
pheno_df <- read_tsv(glue("{PRE}/Br_data/phenotype/pheno.fam"), col_names = FALSE)
names(pheno_df) <- c("FID","IID","bodylen","bmi")

```

### Run LMM Association


```{r lmm function}
## HERE WE USE THE FULL GRM MATRIX AND CALCULATE THE INVERSE OF THE SIGMA MATRIX}
## define lmm association function 
lmmGRM = function(pheno, grm_mat, h2, pred_expr, pheno_id_col=2,pheno_value_cols=6:6,out=NULL)
{
  ## input pheno is a data frame with id column pheno_id_col=1 by default
  ## phenotype values are in pheno_value_cols, 6:6 by default (SCORE column location in plink output), it can have more than one phenotype
  ## but h2 has to be the same, this is useful when running simulations with different h2
  ## call lmmXcan(pheno %>% select(IID,SCORE))
  
  ## format pheno to matrix form
  phenomat <- as.matrix(pheno[,pheno_value_cols])
  rownames(phenomat) <- pheno[[pheno_id_col]]
  
  ## turn pred_expr into matrix with rownames =IID, keep only IIDs in ymat
  exp_mat = as.matrix(pred_expr %>% dplyr::select(-IID))
  rownames(exp_mat) = pred_expr$IID

  ## align pheno and expr matrices
  idlist = intersect(rownames(phenomat), rownames(exp_mat))
  
  nsam = length(idlist)
  
    ## number
  num_genes <- ncol(exp_mat)

  print(glue("{nsam} samples, {num_genes} genes used in association test"))
  ## CALCULATE SIGMA
    ID_mat = diag(rep(1,nsam))
  
  Sigma = grm_mat[idlist,idlist] * h2 + (1 - h2) * ID_mat
  
  Sig_eigen = eigen(Sigma)
  rownames(Sig_eigen$vectors) = rownames(Sigma)
  
  isighalf = Sig_eigen$vectors %*% diag( 1 / sqrt(  Sig_eigen$values  ) ) %*% t(Sig_eigen$vectors)
  

  ## perform raw association
  cormat_raw = matrix_lm(phenomat[idlist,, drop = FALSE], exp_mat[idlist,])
  pmat_raw = cor2pval(cormat_raw,nsam)
  colnames(pmat_raw) <- gsub("cor_", "pval_", colnames(pmat_raw))
  
  ## perform corrected association
  cormat_correct = matrix_lm(isighalf%*% phenomat[idlist,, drop = FALSE], isighalf %*% exp_mat[idlist,])
  pmat_correct = cor2pval(cormat_correct,nsam)
  colnames(pmat_correct) <- gsub("cor_", "pval_", colnames(pmat_correct))
  
  # write outputs to file
  if(!is.null(out))
  {
    saveRDS(cormat_correct,file = glue("{out}_cormat_correct.RDS"))
    saveRDS(pmat_correct,  file = glue("{out}_pmat_correct.RDS"))
    saveRDS(cormat_raw,    file = glue("{out}_cormat_raw.RDS"))
    saveRDS(pmat_raw,      file = glue("{out}_pmat_raw.RDS"))
  }
  res = list(
    cormat_correct=cormat_correct, 
    pmat_correct=pmat_correct, 
    cormat_raw=cormat_raw, 
    pmat_raw=pmat_raw)
  return(res)
  
}
```

#### Association Tests using PredictDB models (Elastic Net)

```{r}
trait = "bodylen"
h2 = bodylen_h2
h2se = bodylen_h2+bodylen_se
tissues = c("AC", "IL", "LH", "PL", "VO")
pred_expr
for (tissue in tissues){
  ## read predicted expression
  pred_expr <- read_tsv(glue("{OUTPUT}/Br_bodylen_bmi_{tissue}__predict.txt")) %>% 
  dplyr::select(-FID) %>%  # Remove the FID column
  mutate(IID = str_split(IID, "_", simplify = TRUE)[, 1])  # Keep the first part of IID
  res_h2 <- lmmGRM(pheno_df,grm_mat, h2, pred_expr,pheno_id_col=2, pheno_value_cols=which(colnames(pheno_df)==trait), out=glue("{OUTPUT}/Br_bodylen_bmi_{tissue}"))
}
```

Aggregate p-values across tissues as ACAT p-values.

```{r}
AC_pval = readRDS(glue("{OUTPUT}/Br_bodylen_bmi_AC_pmat_correct.RDS"))
IL_pval = readRDS(glue("{OUTPUT}/Br_bodylen_bmi_IL_pmat_correct.RDS"))
LH_pval = readRDS(glue("{OUTPUT}/Br_bodylen_bmi_LH_pmat_correct.RDS"))
PL_pval = readRDS(glue("{OUTPUT}/Br_bodylen_bmi_PL_pmat_correct.RDS"))
VO_pval = readRDS(glue("{OUTPUT}/Br_bodylen_bmi_VO_pmat_correct.RDS"))

## Get list of common genes across all tissue
genes <- Reduce(intersect, list(rownames(AC_pval), rownames(IL_pval), rownames(LH_pval), rownames(PL_pval), rownames(VO_pval)))
print(glue("{length(genes)} genes with ACAT p-values"))
```

We plot raw and corrected p-values for VO tissue to confirm inflation.

```{r}
#png(glue("{OUTPUT}/bodylen-AC-lmmGRM.png"))
qqunif.compare(res_h2$pmat_raw,res_h2$pmat_correct,main=glue("RatXcan bodylength, VO tissue") )

```


```{r}
pval_df <- data.frame(
  AC = AC_pval[genes,"bodylen"],
  IL = IL_pval[genes,"bodylen"],
  LH = LH_pval[genes,"bodylen"],
  PL = PL_pval[genes,"bodylen"],
  VO = VO_pval[genes,"bodylen"]
)
pval_df$pvalue <- rowMeans(pval_df)
pval_df$geneId =  genes

```

Manhattan plot of ACAT p-values:

```{r manhattan plot}
## here
library(ggrepel)
gg_manhattan <- function(df, titulo="",significance_threshold = 0.05) {
  ## USAGE: gg_manhattan(df,0.05)
  ## df has columns: pvalue, chr (numeric), and start (position)
  ## significance threshold gets divided by the number of tests
  # Calculate cumulative base pair positions
  data_cum <- df %>%
    group_by(chromosome) %>%
    summarise(max_bp = as.numeric(max(start)), .groups = 'drop') %>%
    mutate(bp_add = lag(cumsum(max_bp), default = 0))

  gwas_data <- df %>%
    inner_join(data_cum, by = "chromosome") %>%
    mutate(bp_cum = start + bp_add)

  # Calculate axis labels
  axis_set <- gwas_data %>%
    group_by(chromosome) %>%
    summarize(center = mean(bp_cum), .groups = 'drop')

  # Determine the ylim based on the most significant p-value
  ylim <- gwas_data %>%
    filter(pvalue == min(pvalue)) %>%
    summarise(ylim = abs(floor(log10(pvalue))) + 2) %>%
    pull(ylim)

  # Calculate the genome-wide significance level
  sig <- significance_threshold / nrow(df)

  # Construct the Manhattan plot
  manhattan_plot <- ggplot(gwas_data, aes(x = bp_cum, y = -log10(pvalue), color = as.factor(chromosome), size = -log10(pvalue))) +
    geom_hline(yintercept = -log10(sig), color = "grey40", linetype = "dashed") +
    geom_hline(yintercept = -log10(0.0001), color = "red", linetype = "dashed") +
    geom_point(alpha = 0.75, shape = 19) + # Simplified shape decision for clarity
    geom_label_repel(aes(label = ifelse(pvalue <= sig, geneSymbol, "")), size = 3) +
    ylim(c(0, ylim)) +
    scale_x_continuous(labels = axis_set$chromosome, breaks = axis_set$tss) +
    scale_color_manual(values = rep(c("dodgerblue4", "midnightblue"), length(unique(axis_set$chromosome)))) +
    scale_size_continuous(range = c(0.5, 3)) +
    labs(x = NULL, y = expression(-log[10](italic(p)))) +
    theme_minimal() +
    theme(legend.position = "none",
          panel.border = element_blank(),
          panel.grid.major.x = element_blank(),
          panel.grid.minor.x = element_blank(),
          axis.text.x = element_text(angle = 90, size = 12),
          axis.text.y = element_text(size = 12, vjust = 0),
          axis.title = element_text(size = 20))

  if(titulo !="") manhattan_plot = manhattan_plot + ggtitle(titulo)
  return(manhattan_plot)
}

```

```{r}
gene_annot <- read.delim(glue("{PRE}/rn7.gene.txt"), header = TRUE, sep = "\t", stringsAsFactors = FALSE)[,c(1:5,7)]
res <- inner_join(pval_df, gene_annot, by="geneId")

```

There are no Bonferroni significant genes, but these are lowest p-values:

```{r}
print(res %>% arrange(pvalue) %>% head(n=20)) %>% dplyr::select(geneSymbol, pvalue)
```

```{r}
gg <- gg_manhattan(res, titulo="RatXcan Bodylength (5 tissue aggregate)")
# ggsave(glue("{OUTPUT}/bodylen-manhattan-p_AC.png"))
print(gg)
```

#### Association using Fast Enformer

```{r}
trait = "bodylen"
h2 = bodylen_h2
h2se = bodylen_h2+bodylen_se
pred_expr <- read_tsv(glue("{OUTPUT}/Br_CAGE_brain__predict.txt")) %>% 
  dplyr::select(-FID) %>%  # Remove the FID column
  mutate(IID = str_split(IID, "_", simplify = TRUE)[, 1])  # Keep the first part of IID
res_h2 <- lmmGRM(pheno_df,grm_mat, h2, pred_expr,pheno_id_col=2, pheno_value_cols=which(colnames(pheno_df)==trait), out=glue("{OUTPUT}/Br_bodylen_bmi_CAGE_brain"))
```


```{r}
#png(glue("{OUTPUT}/bodylen-AC-lmmGRM.png"))
qqunif.compare(res_h2$pmat_raw,res_h2$pmat_correct,main=glue("RatXcan bodylength, CAGE:brain") )

```


```{r}
pval = res_h2$pmat_correct
pval_df = data.frame(geneId = rownames(pval), pvalue = pval[, "bodylen"], row.names = NULL)
res = inner_join(pval_df, gene_annot, by="geneId")

```

There are no Bonferroni significant genes, but these are lowest p-values:

```{r}
print(res %>% arrange(pvalue) %>% head(n=20)) %>% dplyr::select(geneSymbol, pvalue)
```


```{r}
gg <- gg_manhattan(res, titulo="RatXcan Bodylength (CAGE:brain)")
# ggsave(glue("{OUTPUT}/bodylen-manhattan-p_AC.png"))
print(gg)
```





