---
title: Processing 8K HS rat probabilities from qtl2 output
author: Sabrina Mi
date: 08/10/2025
---

I eventually want to organize these probabilities indexed by position in an h5 file, so that given a genomic region, for each individual, we can query probabilities at all snps in the interval. But for now, I just combine the .rds outputs into csvs for each individual, with snp IDs as row names and the 8 founder rats as the columns.

I created a new folder `/eagle/AIHPC4Edu/sabrina/bodylen_bmi_genotype_probabilities` with subdirectories for each chromosome to store csvs for each individual.
Then started up R in a conda environment: `module use /soft/modulefiles; module load conda; conda activate genomics`
## Concatenate Probabilites by Chromosome

```{r global variables}
qtl2_output_dir = "/eagle/AIHPC4Edu/sabrina/scratch/qtl2_outputs/"
output_dir = "/eagle/AIHPC4Edu/sabrina/bodylen_bmi_genotype_probabilities/"
```

```{r functions}
"%&%" = function(a,b) paste(a,b,sep="")
split_by_sample = function(chromosome) {
  chr = as.character(chromosome)
  prob_dir <- qtl2_output_dir %&% "chr" %&% chr %&% "_by_batch/"
  
  batches <- list.files(prob_dir)
  for (file in batches) {
    pr <- readRDS(prob_dir %&% file)[[1]]
    samples <- rownames(pr)
    for (sample in samples) {
      sample_dir = output_dir %&% sample
      if (!file.exists(sample_dir)) {
        dir.create(sample_dir)
      }
      csv = sample_dir %&% "/chr" %&% chr %&% "_probabilities.csv"
      if (!file.exists(csv)) {
        write.csv(t(pr[sample, ,]), csv, quote=FALSE)
      }
    }
  }
}

```

## Write CSVs

```{r run}
for (i in 1:20) {
  cat("Chromosome:", as.character((i)), "\n")
  split_by_sample(i)
}
```
