{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "---\n",
    "title: Preparing pipeline inputs for Enformer predictions centered at canonical TSS\"\n",
    "date: 8/8/23\n",
    "author: Sabrina Mi\n",
    "---"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Write Gene Intervals"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>gene</th>\n",
       "      <th>genename</th>\n",
       "      <th>pred.perf.R2</th>\n",
       "      <th>n.snps.in.model</th>\n",
       "      <th>pred.perf.pval</th>\n",
       "      <th>cor</th>\n",
       "      <th>pred.perf.qval</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>ENSRNOG00000001342</td>\n",
       "      <td>Wdr66</td>\n",
       "      <td>0.844361</td>\n",
       "      <td>3</td>\n",
       "      <td>7.502608e-93</td>\n",
       "      <td>0.918891</td>\n",
       "      <td>1.243114e-89</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>ENSRNOG00000020624</td>\n",
       "      <td>Acadsb</td>\n",
       "      <td>0.829218</td>\n",
       "      <td>3</td>\n",
       "      <td>2.602032e-88</td>\n",
       "      <td>0.910614</td>\n",
       "      <td>2.155665e-85</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>ENSRNOG00000060523</td>\n",
       "      <td>AABR07044362.6</td>\n",
       "      <td>0.820292</td>\n",
       "      <td>8</td>\n",
       "      <td>8.068512e-86</td>\n",
       "      <td>0.905700</td>\n",
       "      <td>4.456263e-83</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>ENSRNOG00000058006</td>\n",
       "      <td>Sncg</td>\n",
       "      <td>0.816705</td>\n",
       "      <td>6</td>\n",
       "      <td>7.472730e-85</td>\n",
       "      <td>0.903717</td>\n",
       "      <td>3.095408e-82</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>ENSRNOG00000060045</td>\n",
       "      <td>Pi4ka</td>\n",
       "      <td>0.815282</td>\n",
       "      <td>3</td>\n",
       "      <td>1.784460e-84</td>\n",
       "      <td>0.902930</td>\n",
       "      <td>5.913374e-82</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                 gene        genename  pred.perf.R2  n.snps.in.model  \\\n",
       "0  ENSRNOG00000001342           Wdr66      0.844361                3   \n",
       "1  ENSRNOG00000020624          Acadsb      0.829218                3   \n",
       "2  ENSRNOG00000060523  AABR07044362.6      0.820292                8   \n",
       "3  ENSRNOG00000058006            Sncg      0.816705                6   \n",
       "4  ENSRNOG00000060045           Pi4ka      0.815282                3   \n",
       "\n",
       "   pred.perf.pval       cor  pred.perf.qval  \n",
       "0    7.502608e-93  0.918891    1.243114e-89  \n",
       "1    2.602032e-88  0.910614    2.155665e-85  \n",
       "2    8.068512e-86  0.905700    4.456263e-83  \n",
       "3    7.472730e-85  0.903717    3.095408e-82  \n",
       "4    1.784460e-84  0.902930    5.913374e-82  "
      ]
     },
     "execution_count": 2,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "import pandas as pd\n",
    "\n",
    "model_genes = pd.read_csv(\"highestR2genes.csv\", header=0)\n",
    "model_genes.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/tmp/ipykernel_164253/1218160640.py:2: DtypeWarning: Columns (2) have mixed types. Specify dtype option on import or set low_memory=False.\n",
      "  annot_df = pd.read_csv('/home/s1mi/enformer_rat_data/rn7.gene.txt', sep ='\\t',  index_col='geneId')\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>geneSymbol</th>\n",
       "      <th>chromosome</th>\n",
       "      <th>start</th>\n",
       "      <th>end</th>\n",
       "      <th>strand</th>\n",
       "      <th>tss</th>\n",
       "      <th>description</th>\n",
       "      <th>hasEqtl</th>\n",
       "      <th>expr_BLA</th>\n",
       "      <th>expr_Brain</th>\n",
       "      <th>...</th>\n",
       "      <th>sqtl_BLA</th>\n",
       "      <th>sqtl_Brain</th>\n",
       "      <th>sqtl_Eye</th>\n",
       "      <th>sqtl_IL</th>\n",
       "      <th>sqtl_LHb</th>\n",
       "      <th>sqtl_NAcc</th>\n",
       "      <th>sqtl_NAcc2</th>\n",
       "      <th>sqtl_OFC</th>\n",
       "      <th>sqtl_PL</th>\n",
       "      <th>sqtl_PL2</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>geneId</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>ENSRNOG00000066169</th>\n",
       "      <td>ENSRNOG00000066169</td>\n",
       "      <td>1</td>\n",
       "      <td>36112690</td>\n",
       "      <td>36122387</td>\n",
       "      <td>-</td>\n",
       "      <td>36122387</td>\n",
       "      <td>NaN</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>...</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSRNOG00000070168</th>\n",
       "      <td>Olr56</td>\n",
       "      <td>1</td>\n",
       "      <td>157231467</td>\n",
       "      <td>157232417</td>\n",
       "      <td>+</td>\n",
       "      <td>157231467</td>\n",
       "      <td>olfactory receptor family 51 subfamily F membe...</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>...</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSRNOG00000070901</th>\n",
       "      <td>Irgq</td>\n",
       "      <td>1</td>\n",
       "      <td>80123925</td>\n",
       "      <td>80131881</td>\n",
       "      <td>+</td>\n",
       "      <td>80123925</td>\n",
       "      <td>immunity-related GTPase Q</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>...</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>3 rows Ã— 68 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                            geneSymbol chromosome      start        end  \\\n",
       "geneId                                                                    \n",
       "ENSRNOG00000066169  ENSRNOG00000066169          1   36112690   36122387   \n",
       "ENSRNOG00000070168               Olr56          1  157231467  157232417   \n",
       "ENSRNOG00000070901                Irgq          1   80123925   80131881   \n",
       "\n",
       "                   strand        tss  \\\n",
       "geneId                                 \n",
       "ENSRNOG00000066169      -   36122387   \n",
       "ENSRNOG00000070168      +  157231467   \n",
       "ENSRNOG00000070901      +   80123925   \n",
       "\n",
       "                                                          description  \\\n",
       "geneId                                                                  \n",
       "ENSRNOG00000066169                                                NaN   \n",
       "ENSRNOG00000070168  olfactory receptor family 51 subfamily F membe...   \n",
       "ENSRNOG00000070901                          immunity-related GTPase Q   \n",
       "\n",
       "                    hasEqtl  expr_BLA  expr_Brain  ...  sqtl_BLA  sqtl_Brain  \\\n",
       "geneId                                             ...                         \n",
       "ENSRNOG00000066169     True      True        True  ...     False       False   \n",
       "ENSRNOG00000070168     True     False        True  ...     False       False   \n",
       "ENSRNOG00000070901     True      True        True  ...     False       False   \n",
       "\n",
       "                    sqtl_Eye  sqtl_IL  sqtl_LHb  sqtl_NAcc  sqtl_NAcc2  \\\n",
       "geneId                                                                   \n",
       "ENSRNOG00000066169     False    False     False      False       False   \n",
       "ENSRNOG00000070168     False    False     False      False       False   \n",
       "ENSRNOG00000070901     False    False     False      False       False   \n",
       "\n",
       "                    sqtl_OFC  sqtl_PL  sqtl_PL2  \n",
       "geneId                                           \n",
       "ENSRNOG00000066169     False    False     False  \n",
       "ENSRNOG00000070168     False    False     False  \n",
       "ENSRNOG00000070901     False    False     False  \n",
       "\n",
       "[3 rows x 68 columns]"
      ]
     },
     "execution_count": 3,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# gene annotation\n",
    "annot_df = pd.read_csv('/home/s1mi/enformer_rat_data/rn7.gene.txt', sep ='\\t',  index_col='geneId')\n",
    "annot_df.head(3)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 33,
   "metadata": {},
   "outputs": [],
   "source": [
    "def write_intervals(gene_list, file):\n",
    "    with open(file, 'a') as f:\n",
    "        for gene in gene_list:\n",
    "            gene_annot = annot_df.loc[gene]\n",
    "            tss = gene_annot['tss']\n",
    "            start = tss - 57344\n",
    "            end = tss + 57344\n",
    "            interval = f\"chr{gene_annot['chromosome']}_{start}_{end}\"\n",
    "            f.write(interval + '\\n')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [],
   "source": [
    "gene_list = [gene for gene in model_genes['gene'][:10] if gene in annot_df.index]\n",
    "with open(\"gene_list.txt\", \"w\") as f:\n",
    "    f.write(\"\\n\".join(gene_list))\n",
    "write_intervals(gene_list, \"metadata/intervals.txt\")\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Write Individuals List"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 38,
   "metadata": {},
   "outputs": [],
   "source": [
    "!bcftools query -l /home/s1mi/enformer_rat_data/Brain.rn7.vcf.gz > metadata/individuals.txt"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "I added these parameters to the config file: `/home/s1mi/Github/deep-learning-in-genomics/posts/running-enformer-on-rat-genes-at-TSS/run_on_polaris_personalized.json`"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Run Enformer\n",
    "\n",
    "```\n",
    "module load conda\n",
    "\n",
    "conda activate /lus/grand/projects/TFXcan/imlab/shared/software/conda_envs/enformer-predict-tools\n",
    "\n",
    "cd /home/s1mi/Github/shared_pipelines/enformer_pipeline\n",
    "python3 scripts/enformer_predict.py --parameters /home/s1mi/Github/deep-learning-in-genomics/posts/running-enformer-on-rat-genes-at-TSS/run_on_polaris_personalized.json\n",
    "\n",
    "```"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Concatenate Predictions"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 41,
   "metadata": {},
   "outputs": [],
   "source": [
    "import os\n",
    "import h5py\n",
    "\n",
    "def concatenate_predictions(predictions_dir, output_path):\n",
    "    with h5py.File(output_path, \"w\") as output_file:\n",
    "        individuals = os.listdir(predictions_dir)\n",
    "        for individual in individuals:\n",
    "            group = output_file.create_group(individual)\n",
    "\n",
    "            dir = os.path.join(predictions_dir, individual, \"haplotype1\")\n",
    "            filenames = os.listdir(dir)\n",
    "\n",
    "            for file in filenames:\n",
    "                interval = file.replace(\"_predictions.h5\", \"\")\n",
    "                with h5py.File(os.path.join(dir, file), \"r\") as input_file:\n",
    "                    group.create_dataset(interval, data=input_file[interval][()], shape=(896,5313))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 42,
   "metadata": {},
   "outputs": [],
   "source": [
    "\n",
    "predictions_dir=\"/home/s1mi/Github/deep-learning-in-genomics/posts/running-enformer-on-rat-genes-at-TSS/predictions_folder/personalized_Br_genes/predictions_2023-08-08/enformer_predictions\"\n",
    "concatenate_predictions(predictions_dir, \"Br_2023-08-08_predictions.h5\")"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "base",
   "language": "python",
   "name": "base"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.8.13"
  },
  "orig_nbformat": 4
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
